<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Lambda Moses" />

<meta name="date" content="2019-06-22" />

<title>Processing kallisto bus Output (10x v2 chemistry)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BUS Notebooks R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BUStools/BUS_notebooks_R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Processing kallisto bus Output (10x v2 chemistry)</h1>
<h4 class="author">Lambda Moses</h4>
<h4 class="date">2019-06-22</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-06-22
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>BUS_notebooks_R/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181214code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181214)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181214code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181214)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtree9fa20ae42471ca2c414164c7a1d726263dfed52btargetblank9fa20aea"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/BUStools/BUS_notebooks_R/tree/9fa20ae42471ca2c414164c7a1d726263dfed52b" target="_blank">9fa20ae</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtree9fa20ae42471ca2c414164c7a1d726263dfed52btargetblank9fa20aea" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    BUS_notebooks_R.Rproj
    Ignored:    data/fastqs/
    Ignored:    data/hgmm_1k_fastqs.tar
    Ignored:    data/hs_cdna.fa.gz
    Ignored:    data/mm_cdna.fa.gz
    Ignored:    data/neuron_10k_fastqs.tar
    Ignored:    data/whitelist_v2.txt
    Ignored:    output/hs_mm_tr_index.idx
    Ignored:    output/hs_tr_index.idx
    Ignored:    output/mm_tr_index.idx
    Ignored:    tmp/

Untracked files:
    Untracked:  output/tr2g_hgmm.tsv

Unstaged changes:
    Modified:   README.md
    Modified:   analysis/10xv3.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/10xv2.html" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/dba27cc4899dd41ffacf23c3c5307ce082466347/analysis/10xv2.Rmd" target="_blank">dba27cc</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Adapted notebook to new version of bustools
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/analysis/10xv2.Rmd" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
<td>
Adapted old notebooks to new version of BUSpaRse
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/10xv2.html" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
<td>
Adapted old notebooks to new version of BUSpaRse
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/c465ce10ae3051256899d4c085684a715c1fc1bf/docs/10xv2.html" target="_blank">c465ce1</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/f22beac3e993d1f097eb2d05c4a0664355921fad/analysis/10xv2.Rmd" target="_blank">f22beac</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Added CellRanger whiltelist clarification
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/000215e50f086ad1c9a3e52c3f47b91bcb1592ad/analysis/10xv2.Rmd" target="_blank">000215e</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Clarified git cloning this repo and resolved swapped code chunks for output.sorted.bus
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/71513e576f231d44fffa4279137aaa6a6d8ff308/docs/10xv2.html" target="_blank">71513e5</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Good site with figures and reproducibility metrics
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/95f29514e0449e913e4c16d3f9b6c470e373b44f/analysis/10xv2.Rmd" target="_blank">95f2951</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Added head of bus file
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/71849682375cab60ab7212e22d9384bb91e7494e/docs/10xv2.html" target="_blank">7184968</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/analysis/10xv2.Rmd" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Finally saved the figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/10xv2.html" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Finally saved the figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/45c5d4aa55bf7ae465dac70a4351f292d262b8f4/docs/10xv2.html" target="_blank">45c5d4a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/92aa9159e1d9e192d401e974d8551a06bfc52c72/analysis/10xv2.Rmd" target="_blank">92aa915</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Named chunks with images
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/87d15f5ac8538c38c74af4f0763f79887c2a5e4d/docs/10xv2.html" target="_blank">87d15f5</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8d9fe9a5ef4f9ee3a3123a4db61029831dacc2e3/analysis/10xv2.Rmd" target="_blank">8d9fe9a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
More detailed explanations for kallisto bus workshop
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/b7b21a0eab033a957a93102ee2222a1d1139d1dd/docs/10xv2.html" target="_blank">b7b21a0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/c7ecb569798bcd9bc6bd40eb052540dc65e39d71/analysis/10xv2.Rmd" target="_blank">c7ecb56</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-02
</td>
<td>
Added BUSpaRse installation link and note
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/a0becdf08662c24bc271e100ba33d3ca88082c49/docs/10xv2.html" target="_blank">a0becdf</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/93c105377b8a2e1f59368d09967380c9b1c53b9f/docs/10xv2.html" target="_blank">93c1053</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-19
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9731fd60718b738ccfc163b04bffe95981fdce6c/analysis/10xv2.Rmd" target="_blank">9731fd6</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-19
</td>
<td>
Corrected typo
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/85a5770b53befbfc968af9e488b546e123a8eab6/docs/10xv2.html" target="_blank">85a5770</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/ca1d6ce5bd620c25609b92c6d098229c1b713bb6/analysis/10xv2.Rmd" target="_blank">ca1d6ce</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Clean up
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/6276894f775df80018e05467e79c54a750722e4f/docs/10xv2.html" target="_blank">6276894</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/db70187594c25264592142e5cd44901a6f8cc795/docs/10xv2.html" target="_blank">db70187</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/b67fc92f091c0b1267328340fe130a14f15f61cd/analysis/10xv2.Rmd" target="_blank">b67fc92</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Added 10xv3 notebook and further elaboration to 10xv2 notebook
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/fff442dd013fb607b3765ce055b06d615421d3da/docs/10xv2.html" target="_blank">fff442d</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/695e2021b83db11959535978b91ec1cdb10292aa/analysis/10xv2.Rmd" target="_blank">695e202</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Changed name to BUSpaRse
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/09a56c1792b414c8e19e99adbaa5845bcb7e7387/docs/10xv2.html" target="_blank">09a56c1</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/d288e19e9b36268288537c66f056598990d32062/analysis/10xv2.Rmd" target="_blank">d288e19</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Cache rather than skip evaluation
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/30c9aa31c60fd01ec97934196bdcaad38e1f920e/docs/10xv2.html" target="_blank">30c9aa3</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/fd3d5ae03b7336d210a1e59af3b2b99f20a6c787/analysis/10xv2.Rmd" target="_blank">fd3d5ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Don’t collapse output
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/2da3b350a1686dc06961608b49e193f1995d74fa/docs/10xv2.html" target="_blank">2da3b35</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/1fc3e91f1f693eef956c8117a016e06851a470f8/analysis/10xv2.Rmd" target="_blank">1fc3e91</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Publish 10xv2 notebook
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/074d55f43942e807d167dab5636ec5bb8a1559fd/analysis/10xv2.Rmd" target="_blank">074d55f</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Initial commit, already with 10xv2 notebook
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>In this vignette, we process fastq data from scRNA-seq (10x v2 chemistry) with to make a sparse matrix that can be used in downstream analysis with command line tools <code>kallisto</code> and <code>bustools</code>, as described in the <a href="https://www.biorxiv.org/content/10.1101/673285v1"><code>kallisto bus</code> paper</a>. Then we will start a standard downstream analysis with <code>Seurat</code>.</p>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>If you would like to rerun this notebook, you can git clone <a href="https://github.com/BUStools/BUS_notebooks_R">this repository</a> or directly download <a href="https://github.com/BUStools/BUS_notebooks_R/blob/master/analysis/10xv2.Rmd">this notebook from GitHub</a>.</p>
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<p>The primary analysis section of this notebook demonstrates the use of command line tools <code>kallisto</code> and <code>bustools</code>. Please use <code>kallisto</code> &gt;= 0.45, whose binary can be downloaded <a href="https://pachterlab.github.io/kallisto/download">here</a>. The binary of <code>bustools</code> can be found <a href="https://github.com/BUStools/bustools/releases">here</a>.</p>
<p>After you download the binary, you should decompress the file (if it is <code>tar.gz</code>) with <code>tar -xzvf file.tar.gz</code> in the <code>bash</code> terminal, and add the directory containing the binary to <code>PATH</code> by <code>export PATH=$PATH:/foo/bar</code>, where <code>/foo/bar</code> here is the directory of interest. Then you can directly invoke the binary on the command line as we will do in this notebook.</p>
<p>We will be using the R packages below. <code>BUSpaRse</code> is not yet on CRAN or Bioconductor. For Mac users, see the <a href="https://github.com/BUStools/BUSpaRse">installation note for <code>BUSpaRse</code></a>. <code>BUSpaRse</code> will be used to generate the transcript to gene file for <code>bustools</code> and to read output of <code>bustools</code> into R. We will also use <code>Seurat</code> version 3 which is now on CRAN.</p>
<pre class="r"><code># Install devtools if it&#39;s not already installed
if (!require(devtools)) {
  install.packages(&quot;devtools&quot;)
}
# Install from GitHub
devtools::install_github(&quot;BUStools/BUSpaRse&quot;)</code></pre>
<p>The package <code>DropletUtils</code> will be used to estimate the number of real cells as opposed to empty droplets. It’s on Bioconductor, and here is how it should be installed:</p>
<pre class="r"><code>if (!require(BiocManager)) {
  install.packages(&quot;BiocManager&quot;)
}
BiocManager::install(&quot;DropletUtils&quot;)</code></pre>
<p>The other R packages below are on CRAN, and can be installed with <code>install.packages</code>.</p>
<pre class="r"><code>library(BUSpaRse)
library(Seurat)
library(tidyverse)
library(DropletUtils)
library(Matrix)
theme_set(theme_bw())</code></pre>
</div>
<div id="download-data" class="section level2">
<h2>Download data</h2>
<p>The data set we are using here is 1k 1:1 Mixture of Fresh Frozen Human (HEK293T) and Mouse (NIH3T3) Cells from the 10x website. First, we download the fastq files (6.34 GB).</p>
<pre class="r"><code>if (!file.exists(&quot;./data/hgmm_1k_fastqs.tar&quot;)) {
  download.file(&quot;http://cf.10xgenomics.com/samples/cell-exp/2.1.0/hgmm_1k/hgmm_1k_fastqs.tar&quot;, destfile = &quot;./data/hgmm_1k_fastqs.tar&quot;, quiet = TRUE)
}</code></pre>
<p>Then untar this file</p>
<pre class="bash"><code>cd ./data
tar -xvf ./hgmm_1k_fastqs.tar</code></pre>
<pre><code>#&gt; fastqs/
#&gt; fastqs/hgmm_1k_S1_L001_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L001_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L001_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L002_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L002_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L002_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L003_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L003_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L003_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L004_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L004_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L004_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L005_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L005_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L005_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L006_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L006_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L006_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L007_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L007_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L007_R2_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L008_I1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L008_R1_001.fastq.gz
#&gt; fastqs/hgmm_1k_S1_L008_R2_001.fastq.gz</code></pre>
</div>
</div>
<div id="generate-the-gene-count-matrix" class="section level1">
<h1>Generate the gene count matrix</h1>
<div id="build-the-kallisto-index" class="section level2">
<h2>Build the <code>kallisto</code> index</h2>
<p>Here we use <a href="https://pachterlab.github.io/kallisto/about">kallisto</a> to pseudoalign the reads to the transcriptome and then to create the <code>bus</code> file to be converted to a sparse matrix. The first step is to build an index of the transcriptome. This data set has both human and mouse cells, so we need both human and mouse transcriptomes. The transcriptomes downloaded here are from Ensembl version 94, released in October 2018.</p>
<pre class="r"><code># Human transcriptome
if (!file.exists(&quot;./data/hs_cdna.fa.gz&quot;)) {
  download.file(&quot;ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz&quot;, &quot;./data/hs_cdna.fa.gz&quot;, method = &quot;wget&quot;, quiet = TRUE)
}
# Mouse transcriptome
if (!file.exists(&quot;./data/mm_cdna.fa.gz&quot;)) {
  download.file(&quot;ftp://ftp.ensembl.org/pub/release-94/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz&quot;, &quot;./data/mm_cdna.fa.gz&quot;, method = &quot;wget&quot;, quiet = TRUE)
}</code></pre>
<pre class="bash"><code># This chunk is in bash
kallisto version</code></pre>
<pre><code>#&gt; kallisto, version 0.46.0</code></pre>
<p>Actually, we don’t need to unzip the fasta files</p>
<pre class="r"><code>if (!file.exists(&quot;./output/hs_mm_tr_index.idx&quot;)) {
  system(&quot;kallisto index -i ./output/hs_mm_tr_index.idx ./data/hs_cdna.fa.gz ./data/mm_cdna.fa.gz&quot;)
}</code></pre>
</div>
<div id="run-kallisto-bus" class="section level2">
<h2>Run <code>kallisto bus</code></h2>
<p>Here we will generate the <code>bus</code> file. Here <code>bus</code> stands for <em>B</em>arbode, <em>U</em>MI, <em>S</em>et (i.e. equivalent class). In text form, it is a table whose first column is the barcode. The second column is the UMI that are associated with the barcode. The third column is the index of the equivalence class reads with the UMI maps to (equivalence class will be explained later). The fourth column is count of reads with this barcode, UMI, and equivalence class combination, which is ignored as one UMI should stand for one molecule. See <a href="https://doi.org/10.1093/bioinformatics/btz279">this paper</a> for more detail.</p>
<p>These are the technologies supported by <code>kallisto bus</code>:</p>
<pre class="r"><code>system(&quot;kallisto bus --list&quot;, intern = TRUE)</code></pre>
<pre><code>#&gt; Warning in system(&quot;kallisto bus --list&quot;, intern = TRUE): running command
#&gt; &#39;kallisto bus --list&#39; had status 1</code></pre>
<pre><code>#&gt;  [1] &quot;List of supported single-cell technologies&quot;
#&gt;  [2] &quot;&quot;                                          
#&gt;  [3] &quot;short name       description&quot;              
#&gt;  [4] &quot;----------       -----------&quot;              
#&gt;  [5] &quot;10xv1            10x version 1 chemistry&quot;  
#&gt;  [6] &quot;10xv2            10x version 2 chemistry&quot;  
#&gt;  [7] &quot;10xv3            10x version 3 chemistry&quot;  
#&gt;  [8] &quot;CELSeq           CEL-Seq&quot;                  
#&gt;  [9] &quot;CELSeq2          CEL-Seq version 2&quot;        
#&gt; [10] &quot;DropSeq          DropSeq&quot;                  
#&gt; [11] &quot;inDrops          inDrops&quot;                  
#&gt; [12] &quot;SCRBSeq          SCRB-Seq&quot;                 
#&gt; [13] &quot;SureCell         SureCell for ddSEQ&quot;       
#&gt; [14] &quot;&quot;                                          
#&gt; attr(,&quot;status&quot;)
#&gt; [1] 1</code></pre>
<p>Here we have 8 samples. Each sample has 3 files: <code>I1</code> means sample index, <code>R1</code> means barcode and UMI, and <code>R2</code> means the piece of cDNA. The <code>-i</code> argument specifies the index file we just built. The <code>-o</code> argument specifies the output directory. The <code>-x</code> argument specifies the sequencing technology used to generate this data set. The <code>-t</code> argument specifies the number of threads used. I ran this on a server and used 8 threads.</p>
<pre class="bash"><code># This chunk is in bash
cd ./data
kallisto bus -i ../output/hs_mm_tr_index.idx -o ../output/out_hgmm1k -x 10xv2 -t8 \
./fastqs/hgmm_1k_S1_L001_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L001_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L002_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L002_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L003_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L003_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L004_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L004_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L005_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L005_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L006_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L006_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L007_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L007_R2_001.fastq.gz \
./fastqs/hgmm_1k_S1_L008_R1_001.fastq.gz ./fastqs/hgmm_1k_S1_L008_R2_001.fastq.gz</code></pre>
<pre><code>#&gt; 
#&gt; [index] k-mer length: 31
#&gt; [index] number of targets: 302,896
#&gt; [index] number of k-mers: 206,125,466
#&gt; [index] number of equivalence classes: 1,252,306
#&gt; [quant] will process sample 1: ./fastqs/hgmm_1k_S1_L001_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L001_R2_001.fastq.gz
#&gt; [quant] will process sample 2: ./fastqs/hgmm_1k_S1_L002_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L002_R2_001.fastq.gz
#&gt; [quant] will process sample 3: ./fastqs/hgmm_1k_S1_L003_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L003_R2_001.fastq.gz
#&gt; [quant] will process sample 4: ./fastqs/hgmm_1k_S1_L004_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L004_R2_001.fastq.gz
#&gt; [quant] will process sample 5: ./fastqs/hgmm_1k_S1_L005_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L005_R2_001.fastq.gz
#&gt; [quant] will process sample 6: ./fastqs/hgmm_1k_S1_L006_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L006_R2_001.fastq.gz
#&gt; [quant] will process sample 7: ./fastqs/hgmm_1k_S1_L007_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L007_R2_001.fastq.gz
#&gt; [quant] will process sample 8: ./fastqs/hgmm_1k_S1_L008_R1_001.fastq.gz
#&gt;                                ./fastqs/hgmm_1k_S1_L008_R2_001.fastq.gz
#&gt; [quant] finding pseudoalignments for the reads ... done
#&gt; [quant] processed 63,252,296 reads, 52,229,344 reads pseudoaligned</code></pre>
<p>See what the outputs are</p>
<pre class="r"><code>list.files(&quot;./output/out_hgmm1k/&quot;)</code></pre>
<pre><code>#&gt; [1] &quot;matrix.ec&quot;       &quot;output.bus&quot;      &quot;run_info.json&quot;   &quot;transcripts.txt&quot;</code></pre>
</div>
<div id="run-bustools" class="section level2">
<h2>Run <code>BUStools</code></h2>
<div id="map-transcripts-to-genes" class="section level3">
<h3>Map transcripts to genes</h3>
<p>For the sparse matrix, most people are interested in how many UMIs per gene per cell, we here we will quantify this from the <code>bus</code> output, and to do so, we need to find which gene corresponds to each transcript. Remember in the output of <code>kallisto bus</code>, there’s the file <code>transcripts.txt</code>. Those are the transcripts in the transcriptome index.</p>
<p>Remember that we downloaded transcriptome FASTA files from Ensembl just now. In FASTA files, each entry is a sequence with a name. In Ensembl FASTA files, the sequence name has genome annotation of the corresponding sequence, so we can extract transcript IDs and corresponding gene IDs and gene names from there.</p>
<pre class="r"><code>tr2g &lt;- transcript2gene(fasta_file = c(&quot;./data/hs_cdna.fa.gz&quot;, &quot;./data/mm_cdna.fa.gz&quot;),
                        kallisto_out_path = &quot;./output/out_hgmm1k&quot;)</code></pre>
<pre><code>#&gt; Reading FASTA file.
#&gt; Reading FASTA file.</code></pre>
<pre><code>#&gt; Sorting transcripts</code></pre>
<pre class="r"><code>head(tr2g)</code></pre>
<pre><code>#&gt;           transcript              gene gene_name
#&gt; 1: ENST00000632684.1 ENSG00000282431.1     TRBD1
#&gt; 2: ENST00000434970.2 ENSG00000237235.2     TRDD2
#&gt; 3: ENST00000448914.1 ENSG00000228985.1     TRDD3
#&gt; 4: ENST00000415118.1 ENSG00000223997.1     TRDD1
#&gt; 5: ENST00000631435.1 ENSG00000282253.1     TRBD1
#&gt; 6: ENST00000390583.1 ENSG00000211923.1  IGHD3-10</code></pre>
<p><code>bustools</code> requires <code>tr2g</code> to be written into a tab delimited file of a specific format: No headers, first column is transcript ID, and second column is the corresponding gene ID. Transcript IDs must be in the same order as in the <code>kallisto</code> index.</p>
<pre class="r"><code># Write tr2g to format required by bustools
save_tr2g_bustools(tr2g, &quot;./output/tr2g_hgmm.tsv&quot;)</code></pre>
<p>A whitelist that contains all the barcodes known to be present in the kit is provided by 10x and comes with CellRanger. A CellRanger installation is required, though we will not run CellRanger here.</p>
<pre class="bash"><code># Copy v2 chemistry whitelist to working directory
cp ~/cellranger-3.0.2/cellranger-cs/3.0.2/lib/python/cellranger/barcodes/737K-august-2016.txt \
./data/whitelist_v2.txt</code></pre>
<p>Then we’re ready to make the gene count matrix. First, <code>bustools</code> runs barcode error correction on the <code>bus</code> file. Then, the corrected <code>bus</code> file is sorted by barcode, UMI, and equivalence classes. Then the UMIs are counted and the counts are collapsed into gene level. Here the <code>|</code> is pipe in bash, just like the magrittr pipe <code>%&gt;%</code> in R, that pipes the output of one command to the next.</p>
<pre class="bash"><code>mkdir ./output/out_hgmm1k/genecount ./tmp
bustools correct -w ./data/whitelist_v2.txt -p ./output/out_hgmm1k/output.bus | \
bustools sort -T tmp/ -t 4 -p - | \
bustools count -o ./output/out_hgmm1k/genecount/genes -g ./output/tr2g_hgmm.tsv \
-e ./output/out_hgmm1k/matrix.ec -t ./output/out_hgmm1k/transcripts.txt --genecounts -</code></pre>
<pre><code>#&gt; mkdir: cannot create directory ‘./tmp’: File exists
#&gt; Found 737280 barcodes in the whitelist
#&gt; Number of hamming dist 1 barcodes = 20550336
#&gt; Processed 52229344 bus records
#&gt; In whitelist = 50774199
#&gt; Corrected = 348400
#&gt; Uncorrected = 1106745
#&gt; Read in 51122599 number of busrecords</code></pre>
<p>See what the outputs are</p>
<pre class="r"><code>list.files(&quot;./output/out_hgmm1k/genecount&quot;)</code></pre>
<pre><code>#&gt; [1] &quot;genes.barcodes.txt&quot; &quot;genes.genes.txt&quot;    &quot;genes.mtx&quot;</code></pre>
<p>Here we have text files for barcodes and gene names, and an <code>mtx</code> file for the sparse gene count matrix.</p>
</div>
</div>
</div>
<div id="explore-the-data" class="section level1">
<h1>Explore the data</h1>
<p>Now we can load the matrix into R for analysis.</p>
<pre class="r"><code>res_mat &lt;- read_count_output(&quot;./output/out_hgmm1k/genecount&quot;,
                             name = &quot;genes&quot;, tcc = FALSE)</code></pre>
<div id="remove-empty-droplets" class="section level2">
<h2>Remove empty droplets</h2>
<p>Cool, so now we have the sparse matrix. What does it look like?</p>
<pre class="r"><code>dim(res_mat)</code></pre>
<pre><code>#&gt; [1]  76416 362063</code></pre>
<p>The number of genes is as expected for two species. There’re way more cells than we expect, which is about 1000. So what’s going on?</p>
<p>How many UMIs per barcode?</p>
<pre class="r"><code>tot_counts &lt;- Matrix::colSums(res_mat)
summary(tot_counts)</code></pre>
<pre><code>#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#&gt;     0.00     1.00     1.00    76.26     8.00 74534.00</code></pre>
<p>The vast majority of “cells” have only a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated method exist (e.g. see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets"><code>emptyDrops</code> in <code>DropletUtils</code></a>), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don’t have the ground truth.</p>
<pre class="r"><code># Compute barcode rank
bc_rank &lt;- barcodeRanks(res_mat)</code></pre>
<pre class="r"><code>qplot(bc_rank$rank, bc_rank$total, geom = &quot;line&quot;) +
  geom_hline(yintercept = metadata(bc_rank)$knee, color = &quot;blue&quot;, linetype = 2) +
  geom_hline(yintercept = metadata(bc_rank)$inflection, color = &quot;green&quot;, linetype = 2) +
  annotate(&quot;text&quot;, x = 1000, y = 1.5 * c(metadata(bc_rank)$knee, metadata(bc_rank)$inflection),
           label = c(&quot;knee&quot;, &quot;inflection&quot;), color = c(&quot;blue&quot;, &quot;green&quot;)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(x = &quot;Rank&quot;, y = &quot;Total reads&quot;)</code></pre>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous y-axis</code></pre>
<p><img src="figure/10xv2.Rmd/rank-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-rank-1">
Past versions of rank-1.png
</button>
</p>
<div id="fig-rank-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/rank-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv2.Rmd/rank-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/rank-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The inflection point looks like a reasonable number of cells.</p>
<pre class="r"><code># Filter the matrix
res_mat &lt;- res_mat[, tot_counts &gt; metadata(bc_rank)$inflection]
dim(res_mat)</code></pre>
<pre><code>#&gt; [1] 76416  1094</code></pre>
</div>
<div id="cell-species" class="section level2">
<h2>Cell species</h2>
<p>How many cells are from humans and how many from mice? The number of cells with mixed species indicates doublet rate.</p>
<pre class="r"><code>gene_species &lt;- ifelse(str_detect(rownames(res_mat), &quot;^ENSMUSG&quot;), &quot;mouse&quot;, &quot;human&quot;)
mouse_inds &lt;- gene_species == &quot;mouse&quot;
human_inds &lt;- gene_species == &quot;human&quot;
# mark cells as mouse or human
cell_species &lt;- tibble(n_mouse_umi = Matrix::colSums(res_mat[mouse_inds,]),
                       n_human_umi = Matrix::colSums(res_mat[human_inds,]),
                       tot_umi = Matrix::colSums(res_mat),
                       prop_mouse = n_mouse_umi / tot_umi,
                       prop_human = n_human_umi / tot_umi)</code></pre>
<pre class="r"><code># Classify species based on proportion of UMI, with cutoff of 90%
cell_species &lt;- cell_species %&gt;% 
  mutate(species = case_when(
    prop_mouse &gt; 0.9 ~ &quot;mouse&quot;,
    prop_human &gt; 0.9 ~ &quot;human&quot;,
    TRUE ~ &quot;mixed&quot;
  ))</code></pre>
<pre class="r"><code>ggplot(cell_species, aes(n_human_umi, n_mouse_umi, color = species)) +
  geom_point(size = 0.5)</code></pre>
<p><img src="figure/10xv2.Rmd/barn-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-barn-1">
Past versions of barn-1.png
</button>
</p>
<div id="fig-barn-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/barn-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/barn-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Great, looks like the vast majority of cells are not mixed.</p>
<pre class="r"><code>cell_species %&gt;% 
  dplyr::count(species) %&gt;% 
  mutate(proportion = n / ncol(res_mat))</code></pre>
<pre><code>#&gt; # A tibble: 3 x 3
#&gt;   species     n proportion
#&gt;   &lt;chr&gt;   &lt;int&gt;      &lt;dbl&gt;
#&gt; 1 human     566    0.517  
#&gt; 2 mixed       3    0.00274
#&gt; 3 mouse     525    0.480</code></pre>
<p>Great, only about 0.3% of cells here are doublets, which is lower than the ~1% 10x lists. Doublet rate tends to be lower when cell concentration is lower. However, doublets can still be formed with cells from the same species, so the number of mixed species “cells” is only a lower bound of doublet rate.</p>
</div>
<div id="dimension-reduction" class="section level2">
<h2>Dimension reduction</h2>
<pre class="r"><code>seu &lt;- CreateSeuratObject(res_mat, min.cells = 3) %&gt;% 
  NormalizeData(verbose = FALSE) %&gt;% 
  ScaleData(verbose = FALSE) %&gt;% 
  FindVariableFeatures(verbose = FALSE)</code></pre>
<pre class="r"><code># Add species to meta data
seu &lt;- AddMetaData(seu, metadata = cell_species$species, col.name = &quot;species&quot;)</code></pre>
<p>See how number of total counts and number of genes expressed are distributed.</p>
<pre class="r"><code>VlnPlot(seu, c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;), group.by = &quot;species&quot;,
        pt.size = 0.1)</code></pre>
<p><img src="figure/10xv2.Rmd/vln-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-vln-1">
Past versions of vln-1.png
</button>
</p>
<div id="fig-vln-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/vln-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/vln-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another QC plot</p>
<pre class="r"><code>ggplot(seu@meta.data, aes(nCount_RNA, nFeature_RNA, color = species)) +
  geom_point(alpha = 0.7, size = 0.5) +
  labs(x = &quot;Total UMI counts per cell&quot;, y = &quot;Number of genes detected&quot;)</code></pre>
<p><img src="figure/10xv2.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-21-1">
Past versions of unnamed-chunk-21-1.png
</button>
</p>
<div id="fig-unnamed-chunk-21-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/unnamed-chunk-21-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The mixed species doublets do look different from human and mouse cells.</p>
<pre class="r"><code>seu &lt;- RunPCA(seu, verbose = FALSE, npcs = 30)
ElbowPlot(seu, ndims = 30)</code></pre>
<p><img src="figure/10xv2.Rmd/elbow-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-elbow-1">
Past versions of elbow-1.png
</button>
</p>
<div id="fig-elbow-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/elbow-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv2.Rmd/elbow-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/elbow-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>DimPlot(seu, reduction = &quot;pca&quot;, pt.size = 0.5, group.by = &quot;species&quot;)</code></pre>
<p><img src="figure/10xv2.Rmd/pca-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pca-1">
Past versions of pca-1.png
</button>
</p>
<div id="fig-pca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/pca-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv2.Rmd/pca-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/pca-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The first PC separates species, as expected. Also as expected, the doublets are in between human and mouse cells in this plot.</p>
<pre class="r"><code>seu &lt;- RunTSNE(seu, dims = 1:20, check_duplicates = FALSE)
DimPlot(seu, reduction = &quot;tsne&quot;, pt.size = 0.5, group.by = &quot;species&quot;)</code></pre>
<p><img src="figure/10xv2.Rmd/tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-tsne-1">
Past versions of tsne-1.png
</button>
</p>
<div id="fig-tsne-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/9fa20ae42471ca2c414164c7a1d726263dfed52b/docs/figure/10xv2.Rmd/tsne-1.png" target="_blank">9fa20ae</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv2.Rmd/tsne-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv2.Rmd/tsne-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The species separate, and the few doublets form its own cluster, as expected.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>#&gt; R version 3.5.2 (2018-12-20)
#&gt; Platform: x86_64-redhat-linux-gnu (64-bit)
#&gt; Running under: CentOS Linux 7 (Core)
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
#&gt; 
#&gt; locale:
#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#&gt; 
#&gt; attached base packages:
#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
#&gt; [8] methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] Matrix_1.2-15               DropletUtils_1.5.3         
#&gt;  [3] SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0
#&gt;  [5] DelayedArray_0.8.0          BiocParallel_1.16.6        
#&gt;  [7] matrixStats_0.54.0          Biobase_2.42.0             
#&gt;  [9] GenomicRanges_1.34.0        GenomeInfoDb_1.18.2        
#&gt; [11] IRanges_2.16.0              S4Vectors_0.20.1           
#&gt; [13] BiocGenerics_0.28.0         forcats_0.4.0              
#&gt; [15] stringr_1.4.0               dplyr_0.8.1                
#&gt; [17] purrr_0.3.2                 readr_1.3.1                
#&gt; [19] tidyr_0.8.3                 tibble_2.1.3               
#&gt; [21] ggplot2_3.2.0               tidyverse_1.2.1            
#&gt; [23] Seurat_3.0.2                BUSpaRse_0.99.4            
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;   [1] readxl_1.3.1             backports_1.1.4         
#&gt;   [3] workflowr_1.4.0          plyr_1.8.4              
#&gt;   [5] igraph_1.2.4.1           lazyeval_0.2.2          
#&gt;   [7] splines_3.5.2            listenv_0.7.0           
#&gt;   [9] digest_0.6.19            htmltools_0.3.6         
#&gt;  [11] fansi_0.4.0              gdata_2.18.0            
#&gt;  [13] magrittr_1.5             memoise_1.1.0           
#&gt;  [15] cluster_2.0.7-1          ROCR_1.0-7              
#&gt;  [17] limma_3.38.3             globals_0.12.4          
#&gt;  [19] Biostrings_2.50.2        modelr_0.1.4            
#&gt;  [21] RcppParallel_4.4.3       R.utils_2.9.0           
#&gt;  [23] prettyunits_1.0.2        colorspace_1.4-1        
#&gt;  [25] rvest_0.3.4              blob_1.1.1              
#&gt;  [27] ggrepel_0.8.1            haven_2.1.0             
#&gt;  [29] xfun_0.7                 crayon_1.3.4            
#&gt;  [31] RCurl_1.95-4.12          jsonlite_1.6            
#&gt;  [33] zeallot_0.1.0            survival_2.44-1.1       
#&gt;  [35] zoo_1.8-5                ape_5.3                 
#&gt;  [37] glue_1.3.1               gtable_0.3.0            
#&gt;  [39] zlibbioc_1.28.0          XVector_0.22.0          
#&gt;  [41] plyranges_1.2.0          Rhdf5lib_1.4.3          
#&gt;  [43] future.apply_1.3.0       HDF5Array_1.10.1        
#&gt;  [45] scales_1.0.0             edgeR_3.24.3            
#&gt;  [47] DBI_1.0.0                bibtex_0.4.2            
#&gt;  [49] Rcpp_1.0.1               metap_1.1               
#&gt;  [51] viridisLite_0.3.0        progress_1.2.2          
#&gt;  [53] dqrng_0.2.1              reticulate_1.12         
#&gt;  [55] bit_1.1-14               rsvd_1.0.1              
#&gt;  [57] SDMTools_1.1-221.1       tsne_0.1-3              
#&gt;  [59] htmlwidgets_1.3          httr_1.4.0              
#&gt;  [61] gplots_3.0.1.1           RColorBrewer_1.1-2      
#&gt;  [63] ica_1.0-2                pkgconfig_2.0.2         
#&gt;  [65] XML_3.98-1.20            R.methodsS3_1.7.1       
#&gt;  [67] utf8_1.1.4               locfit_1.5-9.1          
#&gt;  [69] labeling_0.3             tidyselect_0.2.5        
#&gt;  [71] rlang_0.3.4              reshape2_1.4.3          
#&gt;  [73] AnnotationDbi_1.44.0     cellranger_1.1.0        
#&gt;  [75] munsell_0.5.0            tools_3.5.2             
#&gt;  [77] cli_1.1.0                generics_0.0.2          
#&gt;  [79] RSQLite_2.1.1            broom_0.5.2             
#&gt;  [81] ggridges_0.5.1           evaluate_0.13           
#&gt;  [83] yaml_2.2.0               npsurv_0.4-0            
#&gt;  [85] knitr_1.22               bit64_0.9-7             
#&gt;  [87] fs_1.3.1                 fitdistrplus_1.0-14     
#&gt;  [89] caTools_1.17.1.2         RANN_2.6.1              
#&gt;  [91] pbapply_1.4-0            future_1.13.0           
#&gt;  [93] nlme_3.1-137             whisker_0.3-2           
#&gt;  [95] R.oo_1.22.0              xml2_1.2.0              
#&gt;  [97] biomaRt_2.38.0           rstudioapi_0.10         
#&gt;  [99] compiler_3.5.2           plotly_4.9.0            
#&gt; [101] png_0.1-7                lsei_1.2-0              
#&gt; [103] stringi_1.4.3            lattice_0.20-38         
#&gt; [105] vctrs_0.1.0              pillar_1.4.1            
#&gt; [107] Rdpack_0.11-0            lmtest_0.9-37           
#&gt; [109] data.table_1.12.2        cowplot_0.9.4           
#&gt; [111] bitops_1.0-6             irlba_2.3.3             
#&gt; [113] gbRd_0.4-11              rtracklayer_1.42.2      
#&gt; [115] R6_2.4.0                 KernSmooth_2.23-15      
#&gt; [117] gridExtra_2.3            codetools_0.2-15        
#&gt; [119] MASS_7.3-51.4            gtools_3.8.1            
#&gt; [121] assertthat_0.2.1         rhdf5_2.26.2            
#&gt; [123] rprojroot_1.3-2          withr_2.1.2             
#&gt; [125] GenomicAlignments_1.18.1 sctransform_0.2.0       
#&gt; [127] Rsamtools_1.34.1         GenomeInfoDbData_1.2.0  
#&gt; [129] hms_0.4.2                grid_3.5.2              
#&gt; [131] rmarkdown_1.12           Rtsne_0.15              
#&gt; [133] git2r_0.25.2             lubridate_1.7.4</code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
