<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Lambda Moses" />

<meta name="date" content="2019-06-23" />

<title>1k PBMCs from a Healthy Donor (10x v3 chemistry)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">kallisto | bustools R Notebooks</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BUStools/BUS_notebooks_R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">1k PBMCs from a Healthy Donor (10x v3 chemistry)</h1>
<h4 class="author">Lambda Moses</h4>
<h4 class="date">2019-06-23</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-06-23
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>BUS_notebooks_R/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181214code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181214)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181214code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181214)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtreeced765491640d968e2ac7d6fd01458ae8d3fe17ctargetblankced7654a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/BUStools/BUS_notebooks_R/tree/ced765491640d968e2ac7d6fd01458ae8d3fe17c" target="_blank">ced7654</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtreeced765491640d968e2ac7d6fd01458ae8d3fe17ctargetblankced7654a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    BUS_notebooks_R.Rproj
    Ignored:    data/whitelist_v3.txt.gz
    Ignored:    output/out_pbmc1k/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/ced765491640d968e2ac7d6fd01458ae8d3fe17c/analysis/10xv3.Rmd" target="_blank">ced7654</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-23
</td>
<td>
Addded explanation of kallisto output
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/df6e9d6dba99db0113a4da73ced3bea9f36ffad2/docs/10xv3.html" target="_blank">df6e9d6</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/2b6460d84d0123080d8aaf64c329dc0d1792bdaa/analysis/10xv3.Rmd" target="_blank">2b6460d</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Transposed knee plot
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/10xv3.html" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/ac15697975ff6fc72b3ca90a50568df65c87a149/analysis/10xv3.Rmd" target="_blank">ac15697</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Changed 10xv3 notebook to demonstrate TENxBUSData on PBMC data
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/d73779a7f66d238c677c25ed844c9e2dc2feb867/analysis/10xv3.Rmd" target="_blank">d73779a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
<td>
Changed 10xv3 notebook
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/analysis/10xv3.Rmd" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
<td>
Adapted old notebooks to new version of BUSpaRse
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/10xv3.html" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
<td>
Adapted old notebooks to new version of BUSpaRse
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/c465ce10ae3051256899d4c085684a715c1fc1bf/docs/10xv3.html" target="_blank">c465ce1</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/f22beac3e993d1f097eb2d05c4a0664355921fad/analysis/10xv3.Rmd" target="_blank">f22beac</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Added CellRanger whiltelist clarification
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/000215e50f086ad1c9a3e52c3f47b91bcb1592ad/analysis/10xv3.Rmd" target="_blank">000215e</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Clarified git cloning this repo and resolved swapped code chunks for output.sorted.bus
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/71513e576f231d44fffa4279137aaa6a6d8ff308/docs/10xv3.html" target="_blank">71513e5</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Good site with figures and reproducibility metrics
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/6006364fcbd35ab19d13d7c1dbd133357a573512/analysis/10xv3.Rmd" target="_blank">6006364</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Corrected typo
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/95f29514e0449e913e4c16d3f9b6c470e373b44f/analysis/10xv3.Rmd" target="_blank">95f2951</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Added head of bus file
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/71849682375cab60ab7212e22d9384bb91e7494e/docs/10xv3.html" target="_blank">7184968</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/analysis/10xv3.Rmd" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Finally saved the figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/10xv3.html" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Finally saved the figures
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/45c5d4aa55bf7ae465dac70a4351f292d262b8f4/docs/10xv3.html" target="_blank">45c5d4a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/92aa9159e1d9e192d401e974d8551a06bfc52c72/analysis/10xv3.Rmd" target="_blank">92aa915</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Named chunks with images
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/87d15f5ac8538c38c74af4f0763f79887c2a5e4d/docs/10xv3.html" target="_blank">87d15f5</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8d9fe9a5ef4f9ee3a3123a4db61029831dacc2e3/analysis/10xv3.Rmd" target="_blank">8d9fe9a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
<td>
More detailed explanations for kallisto bus workshop
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/b7b21a0eab033a957a93102ee2222a1d1139d1dd/docs/10xv3.html" target="_blank">b7b21a0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/c7ecb569798bcd9bc6bd40eb052540dc65e39d71/analysis/10xv3.Rmd" target="_blank">c7ecb56</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-02
</td>
<td>
Added BUSpaRse installation link and note
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/85a5770b53befbfc968af9e488b546e123a8eab6/docs/10xv3.html" target="_blank">85a5770</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/ca1d6ce5bd620c25609b92c6d098229c1b713bb6/analysis/10xv3.Rmd" target="_blank">ca1d6ce</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Clean up
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/6276894f775df80018e05467e79c54a750722e4f/docs/10xv3.html" target="_blank">6276894</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/166854fe0508a3781bd961e329b843464151201e/docs/10xv3.html" target="_blank">166854f</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/ef5cfc5bd78a314089158309ce9ee482384d549b/analysis/10xv3.Rmd" target="_blank">ef5cfc5</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Corrected stated file size
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/BUStools/BUS_notebooks_R/e0e7d0a7267201f05098eb6245cf9e754646c3de/docs/10xv3.html" target="_blank">e0e7d0a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/aab422a6b0dc03205e25c321268bf0959dd4d69c/analysis/10xv3.Rmd" target="_blank">aab422a</a>
</td>
<td>
Lambda Moses
</td>
<td>
2018-12-14
</td>
<td>
Added 10xv3 notebook
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This vignette demonstrates how to manipulate <code>bus</code> format in R with <code>BUSpaRse</code>. The most recent version of <code>bustools</code> can generate gene count matrices from <code>bus</code> files more efficiently; the purpose of the separate implementation in <code>BUSpaRse</code> is for advanced users to experiment with new ways to collapse UMIs mapped to multiple genes and to adapt <code>bus</code> format to purposes other than single cell RNA-seq. This implementation is intended to facilitate exploration using R/Rcpp, which is easier to work with than C++.</p>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>If you would like to rerun this notebook, you can git clone <a href="https://github.com/BUStools/BUS_notebooks_R">this repository</a> or directly download <a href="https://github.com/BUStools/BUS_notebooks_R/blob/master/analysis/10xv3.Rmd">this notebook from GitHub</a>.</p>
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<p>We will be using the R packages below. <code>BUSpaRse</code> is not yet on CRAN or Bioconductor. For Mac users, see the <a href="https://github.com/BUStools/BUSpaRse">installation note for <code>BUSpaRse</code></a>. The data, which is already in <code>bus</code> format, can be downloaded from the package <code>TENxBUSData</code>. Both <code>TENxBUSData</code> and <code>BUSpaRse</code> have been submitted to Bioconductor for review; the data for <code>TENxBUSData</code> can only be downloaded from Bioconductor devel (version 3.10), which requires R 3.6.0.</p>
<pre class="r"><code># Install devtools if it&#39;s not already installed
if (!require(devtools)) {
  install.packages(&quot;devtools&quot;)
}
# Install from GitHub
devtools::install_github(&quot;BUStools/BUSpaRse&quot;)
devtools::install_github(&quot;BUStools/TENxBUSData&quot;)</code></pre>
<p>The package <code>DropletUtils</code> will be used to estimate the number of real cells as opposed to empty droplets. It’s on Bioconductor, and here is how it should be installed:</p>
<pre class="r"><code>if (!require(BiocManager)) {
  install.packages(&quot;BiocManager&quot;)
}
# Install Bioconductor devel
BiocManager::install(version = &quot;devel&quot;)
BiocManager::install(&quot;DropletUtils&quot;)</code></pre>
<p>The other R packages below are on CRAN, and can be installed with <code>install.packages</code>.</p>
<pre class="r"><code>library(BUSpaRse)
library(TENxBUSData)
library(ggplot2)
library(magrittr)
library(data.table)
library(Seurat)
library(DropletUtils)
library(Matrix)
theme_set(theme_bw())</code></pre>
<p>We will not get into the details of how to make the <code>bus</code> file with <code>kallisto bus</code> and <code>bustools</code>, as the data will be downloaded with <code>TENxBUSData</code>. <code>TENxBUSData</code> provides 5 different datasets, and we will use the PBMC 1k dataset here. The data from <code>TENxBUSData</code> contains the sorted <code>bus</code> file in text format. While the <code>BUSpaRse</code> package converts that text format <code>bus</code> file into gene count matrix, this text file can be loaded into R as a data frame for further exploration.</p>
<pre class="r"><code>fn &lt;- TENxBUSData(&quot;./output&quot;, dataset = &quot;pbmc1k&quot;)</code></pre>
<pre><code>#&gt; The dataset has already been downloaded. It is located in /Users/lambda/BUS_notebooks_R/output/out_pbmc1k</code></pre>
<pre class="r"><code>list.files(fn)</code></pre>
<pre><code>#&gt; [1] &quot;matrix.ec&quot;         &quot;output.sorted&quot;     &quot;output.sorted.txt&quot;
#&gt; [4] &quot;transcripts.txt&quot;</code></pre>
<p>Explaining the output:</p>
<ul>
<li><code>matrix.ec</code>: A text file with two columns. The first column is the 0 based index of equivalence classes. The second column is the set of transcripts (denoted by 0 based index based on order of appearance in the transcriptome fasta file) present in the corresponding equivalence class.</li>
<li><code>output.sorted</code>: The data represented in <code>bus</code> format, sorted by barcode, UMI, and equivalence class. This is a binary, so can’t be read into R with functions like <code>read.table</code>.</li>
<li><code>output.sorted.txt</code>: <code>output.sorted</code> converted into text format, so can be easily read into R for exploration.</li>
<li><code>transcript.txt</code>: A text file with one column, which is the transcripts present in the data, in the same order as in the transcriptome fasta file.</li>
</ul>
</div>
</div>
<div id="sparse-matrix" class="section level1">
<h1>Sparse matrix</h1>
<div id="map-transcripts-to-genes" class="section level2">
<h2>Map transcripts to genes</h2>
<p>For the sparse matrix, most people are interested in how many UMIs per gene per cell, we here we will quantify this from the <code>bus</code> output, and to do so, we need to find which gene corresponds to each transcript. Remember in the output of <code>kallisto bus</code>, there’s the file <code>transcripts.txt</code>. Those are the transcripts in the transcriptome index. Information on which transcript corresponds to which gene can be directly retrieved from Ensembl.</p>
<pre class="r"><code>tr2g &lt;- transcript2gene(species = &quot;Homo sapiens&quot;, 
                        kallisto_out_path = &quot;./output/out_pbmc1k&quot;,
                        ensembl_version = 94)</code></pre>
<pre><code>#&gt; Querying biomart for transcript and gene IDs of Homo sapiens</code></pre>
<pre><code>#&gt; Sorting transcripts</code></pre>
<pre class="r"><code>head(tr2g)</code></pre>
<pre><code>#&gt;           transcript              gene gene_name
#&gt; 1: ENST00000632684.1 ENSG00000282431.1     TRBD1
#&gt; 2: ENST00000434970.2 ENSG00000237235.2     TRDD2
#&gt; 3: ENST00000448914.1 ENSG00000228985.1     TRDD3
#&gt; 4: ENST00000415118.1 ENSG00000223997.1     TRDD1
#&gt; 5: ENST00000631435.1 ENSG00000282253.1     TRBD1
#&gt; 6: ENST00000390583.1 ENSG00000211923.1  IGHD3-10</code></pre>
<p>Alternative ways of getting <code>tr2g</code> have been implemented in the <code>BUSpaRse</code> package. You may use <code>tr2g_ensembl</code> to query Ensembl with biomart to get transcript and gene IDs. If you use this method, then please make sure that the Ensembl version used in the query matches that of the transcriptome. This method is convenient for the user since you only need to input species names, but it can be slow since biomart database query can be slow. You may also use <code>tr2g_gtf</code> for GTF files and <code>tr2g_gff3</code> for GFF3 files, which are more useful for non-model organisms absent from Ensemble. After calling the <code>tr2g_*</code> family of functions, you should sort the transcripts from those functions with <code>sort_tr2g</code> so the transcripts are in the same order as those in the kallisto index. Then the function <code>save_tr2g_bustools</code> can be used to save the <code>tr2g</code> data frame to a text file in the format required by <code>bustools</code>.</p>
</div>
<div id="make-the-sparse-matrix" class="section level2">
<h2>Make the sparse matrix</h2>
<p>For 10x, we do have a file with all valid cell barcodes that comes with CellRanger. You need to install CellRanger to get this file, though you do not need to run CellRanger for this notebook. The whitelist is optional, so if you don’t have one, you may skip the whitelist step and the <code>whitelist</code> argument in the <code>makr_sparse_matrix</code> function.</p>
<pre class="bash"><code># Copy v3 chemistry whitelist to working directory
cp ~/cellranger-3.0.2/cellranger-cs/3.0.2/lib/python/cellranger/barcodes/3M-february-2018.txt.gz \
./data/whitelist_v3.txt.gz</code></pre>
<pre class="r"><code># Read in the whitelist
whitelist_v3 &lt;- fread(&quot;./data/whitelist_v3.txt.gz&quot;, header = FALSE)$V1
length(whitelist_v3)</code></pre>
<pre><code>#&gt; [1] 6794880</code></pre>
<p>That’s an order of magnitude more than the 737K in v2 chemistry.</p>
<p>Now we have everything we need to make the sparse matrix. This function reads in <code>output.sorted.txt</code> line by line and processes them. It does not do barcode correction for now, so the barcode must exactly match those in the whitelist if one is provided. It took 5 to 6 minutes to construct the sparse matrix in the hgmm6k dataset, which has over 280 million lines in <code>output.sorted.txt</code>, which is over 9GB. Here the data set is smaller, and it takes less than a minute.</p>
<p>Note that the arguments <code>est_ncells</code> (estimated number of cells) and <code>est_ngenes</code> (estimated number of genes) are important. With the estimate, this function reserves memory for the data to be added into, reducing the need of reallocation, which will slow the function down. Since the vast majority of “cells” you get in this sparse matrix are empty droplets rather than cells, please put at least 200 times more “cells” than you actually expect in <code>est_ncells</code>.</p>
<p>If you do not have a whitelist of barcodes, then it’s fine; the <code>whitelist</code> argument is optional.</p>
<p>The function <code>make_sparse_matrix</code> can make the gene count matrix and the transcript compatibility count (TCC) matrix at the same time. For the purpose of this notebook, we only generate the gene count matrix. An upcoming notebook will demonstrate some more detailed analysis with a TCC matrix. See <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0970-8">Ntranos et al. 2016</a> for more information about TCC matrices.</p>
<pre class="r"><code>res_mat &lt;- make_sparse_matrix(&quot;./output/out_pbmc1k/output.sorted.txt&quot;,
                              tr2g = tr2g, est_ncells = 3e5,
                              est_ngenes = nrow(tr2g),
                              whitelist = whitelist_v3, TCC = FALSE)</code></pre>
<pre><code>#&gt; Reading matrix.ec
#&gt; Processing ECs
#&gt; Matching genes to ECs
#&gt; Reading data
#&gt; Read 5 million reads
#&gt; Read 10 million reads
#&gt; Read 15 million reads
#&gt; Constructing gene count matrix</code></pre>
<p>The matrix we get has genes in rows and barcode in columns. The row names are the gene IDs (not using human readable gene names since they’re not guaranteed to be unique), and the column names are cell barcodes.</p>
</div>
</div>
<div id="explore-the-data" class="section level1">
<h1>Explore the data</h1>
<div id="remove-empty-droplets" class="section level2">
<h2>Remove empty droplets</h2>
<p>Cool, so now we have the sparse matrix. What does it look like?</p>
<pre class="r"><code>dim(res_mat)</code></pre>
<pre><code>#&gt; [1]  19821 216752</code></pre>
<p>That’s way more cells than we expect, which is about 1000. So what’s going on?</p>
<p>How many UMIs per barcode?</p>
<pre class="r"><code>tot_counts &lt;- Matrix::colSums(res_mat)
summary(tot_counts)</code></pre>
<pre><code>#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#&gt;     1.00     1.00     2.00    46.49     6.00 54175.00</code></pre>
<p>The vast majority of “cells” have only a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated method exist (e.g. see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets"><code>emptyDrops</code> in <code>DropletUtils</code></a>), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don’t have the ground truth.</p>
<pre class="r"><code># Compute barcode rank
bc_rank &lt;- barcodeRanks(res_mat)</code></pre>
<pre class="r"><code>qplot(bc_rank$total, bc_rank$rank, geom = &quot;line&quot;) +
  geom_vline(xintercept = metadata(bc_rank)$knee, color = &quot;blue&quot;, linetype = 2) +
  geom_vline(xintercept = metadata(bc_rank)$inflection, color = &quot;green&quot;, linetype = 2) +
  annotate(&quot;text&quot;, y = 1000, x = 1.5 * c(metadata(bc_rank)$knee, metadata(bc_rank)$inflection),
           label = c(&quot;knee&quot;, &quot;inflection&quot;), color = c(&quot;blue&quot;, &quot;green&quot;)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(y = &quot;Barcode rank&quot;, x = &quot;Total UMI count&quot;)</code></pre>
<p><img src="figure/10xv3.Rmd/rank-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-rank-1">
Past versions of rank-1.png
</button>
</p>
<div id="fig-rank-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/df6e9d6dba99db0113a4da73ced3bea9f36ffad2/docs/figure/10xv3.Rmd/rank-1.png" target="_blank">df6e9d6</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/rank-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv3.Rmd/rank-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv3.Rmd/rank-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The inflection point looks like a reasonable number of cells.</p>
<pre class="r"><code># Filter the matrix
res_mat &lt;- res_mat[, tot_counts &gt; metadata(bc_rank)$inflection]
dim(res_mat)</code></pre>
<pre><code>#&gt; [1] 19821  1169</code></pre>
</div>
<div id="dimension-reduction" class="section level2">
<h2>Dimension reduction</h2>
<pre class="r"><code>seu &lt;- CreateSeuratObject(res_mat, min.cells = 3) %&gt;% 
  NormalizeData(verbose = FALSE) %&gt;% 
  ScaleData(verbose = FALSE) %&gt;% 
  FindVariableFeatures(verbose = FALSE)</code></pre>
<p>See how number of total counts and number of genes expressed are distributed.</p>
<pre class="r"><code>VlnPlot(seu, c(&quot;nCount_RNA&quot;, &quot;nFeature_RNA&quot;), pt.size = 0.1)</code></pre>
<p><img src="figure/10xv3.Rmd/vln-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-vln-1">
Past versions of vln-1.png
</button>
</p>
<div id="fig-vln-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/vln-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv3.Rmd/vln-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Another QC plot</p>
<pre class="r"><code>ggplot(seu@meta.data, aes(nCount_RNA, nFeature_RNA)) +
  geom_point(alpha = 0.7, size = 0.5) +
  labs(x = &quot;Total UMI counts per cell&quot;, y = &quot;Number of genes detected&quot;)</code></pre>
<p><img src="figure/10xv3.Rmd/unnamed-chunk-13-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-13-1">
Past versions of unnamed-chunk-13-1.png
</button>
</p>
<div id="fig-unnamed-chunk-13-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/unnamed-chunk-13-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>seu &lt;- RunPCA(seu, verbose = FALSE, npcs = 30)
ElbowPlot(seu, ndims = 30)</code></pre>
<p><img src="figure/10xv3.Rmd/elbow-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-elbow-1">
Past versions of elbow-1.png
</button>
</p>
<div id="fig-elbow-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/elbow-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv3.Rmd/elbow-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv3.Rmd/elbow-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can do Leiden clustering. Leiden is an improvement over Louvain that guarantees that clusters are well-connected on the k nearest neighbor graph.</p>
<pre class="r"><code># Leiden clustering
seu &lt;- FindNeighbors(seu)</code></pre>
<pre><code>#&gt; Computing nearest neighbor graph</code></pre>
<pre><code>#&gt; Computing SNN</code></pre>
<pre class="r"><code>seu &lt;- FindClusters(seu, algorithm = 4)</code></pre>
<pre class="r"><code>DimPlot(seu, reduction = &quot;pca&quot;, pt.size = 0.5)</code></pre>
<p><img src="figure/10xv3.Rmd/pca-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-pca-1">
Past versions of pca-1.png
</button>
</p>
<div id="fig-pca-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/df6e9d6dba99db0113a4da73ced3bea9f36ffad2/docs/figure/10xv3.Rmd/pca-1.png" target="_blank">df6e9d6</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/pca-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv3.Rmd/pca-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv3.Rmd/pca-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>seu &lt;- RunTSNE(seu, dims = 1:20, check_duplicates = FALSE)
DimPlot(seu, reduction = &quot;tsne&quot;, pt.size = 0.5)</code></pre>
<p><img src="figure/10xv3.Rmd/tsne-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-tsne-1">
Past versions of tsne-1.png
</button>
</p>
<div id="fig-tsne-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/df6e9d6dba99db0113a4da73ced3bea9f36ffad2/docs/figure/10xv3.Rmd/tsne-1.png" target="_blank">df6e9d6</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3642b69dd6fea71a98430c4e0bc3201fee7684ab/docs/figure/10xv3.Rmd/tsne-1.png" target="_blank">3642b69</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-06-22
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/06e98693d04189ba4f1b7c49d9466b6663e2e590/docs/figure/10xv3.Rmd/tsne-1.png" target="_blank">06e9869</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-03-01
</td>
</tr>
<tr>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/8cd81e033d3bf2b02d3287043329aea3b3e60d8e/docs/figure/10xv3.Rmd/tsne-1.png" target="_blank">8cd81e0</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-02-14
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>#&gt; R version 3.6.0 (2019-04-26)
#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)
#&gt; Running under: macOS Mojave 10.14.5
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
#&gt; 
#&gt; locale:
#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#&gt; 
#&gt; attached base packages:
#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
#&gt; [8] methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] Matrix_1.2-17               DropletUtils_1.5.4         
#&gt;  [3] SingleCellExperiment_1.7.0  SummarizedExperiment_1.15.4
#&gt;  [5] DelayedArray_0.11.2         BiocParallel_1.19.0        
#&gt;  [7] matrixStats_0.54.0          Biobase_2.45.0             
#&gt;  [9] GenomicRanges_1.37.12       GenomeInfoDb_1.21.1        
#&gt; [11] IRanges_2.19.10             S4Vectors_0.23.13          
#&gt; [13] BiocGenerics_0.31.4         Seurat_3.0.2               
#&gt; [15] data.table_1.12.2           magrittr_1.5               
#&gt; [17] ggplot2_3.2.0               TENxBUSData_0.99.2         
#&gt; [19] BUSpaRse_0.99.4            
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;   [1] backports_1.1.4               AnnotationHub_2.17.3         
#&gt;   [3] workflowr_1.4.0               BiocFileCache_1.9.1          
#&gt;   [5] plyr_1.8.4                    igraph_1.2.4.1               
#&gt;   [7] lazyeval_0.2.2                splines_3.6.0                
#&gt;   [9] listenv_0.7.0                 digest_0.6.19                
#&gt;  [11] htmltools_0.3.6               gdata_2.18.0                 
#&gt;  [13] memoise_1.1.0                 cluster_2.1.0                
#&gt;  [15] ROCR_1.0-7                    limma_3.41.5                 
#&gt;  [17] globals_0.12.4                Biostrings_2.53.0            
#&gt;  [19] RcppParallel_4.4.3            R.utils_2.9.0                
#&gt;  [21] prettyunits_1.0.2             colorspace_1.4-1             
#&gt;  [23] blob_1.1.1                    rappdirs_0.3.1               
#&gt;  [25] ggrepel_0.8.1                 xfun_0.7                     
#&gt;  [27] dplyr_0.8.1                   jsonlite_1.6                 
#&gt;  [29] crayon_1.3.4                  RCurl_1.95-4.12              
#&gt;  [31] zeallot_0.1.0                 survival_2.44-1.1            
#&gt;  [33] zoo_1.8-6                     ape_5.3                      
#&gt;  [35] glue_1.3.1                    gtable_0.3.0                 
#&gt;  [37] zlibbioc_1.31.0               XVector_0.25.0               
#&gt;  [39] plyranges_1.5.6               Rhdf5lib_1.7.2               
#&gt;  [41] future.apply_1.3.0            HDF5Array_1.13.2             
#&gt;  [43] scales_1.0.0                  edgeR_3.27.5                 
#&gt;  [45] DBI_1.0.0                     bibtex_0.4.2                 
#&gt;  [47] Rcpp_1.0.1                    metap_1.1                    
#&gt;  [49] viridisLite_0.3.0             xtable_1.8-4                 
#&gt;  [51] progress_1.2.2                dqrng_0.2.1                  
#&gt;  [53] reticulate_1.12               rsvd_1.0.1                   
#&gt;  [55] bit_1.1-14                    SDMTools_1.1-221.1           
#&gt;  [57] tsne_0.1-3                    htmlwidgets_1.3              
#&gt;  [59] httr_1.4.0                    gplots_3.0.1.1               
#&gt;  [61] RColorBrewer_1.1-2            ica_1.0-2                    
#&gt;  [63] pkgconfig_2.0.2               XML_3.98-1.20                
#&gt;  [65] R.methodsS3_1.7.1             dbplyr_1.4.2                 
#&gt;  [67] locfit_1.5-9.1                labeling_0.3                 
#&gt;  [69] reshape2_1.4.3                tidyselect_0.2.5             
#&gt;  [71] rlang_0.3.4                   later_0.8.0                  
#&gt;  [73] AnnotationDbi_1.47.0          munsell_0.5.0                
#&gt;  [75] tools_3.6.0                   RSQLite_2.1.1                
#&gt;  [77] ExperimentHub_1.11.1          ggridges_0.5.1               
#&gt;  [79] evaluate_0.14                 stringr_1.4.0                
#&gt;  [81] yaml_2.2.0                    npsurv_0.4-0                 
#&gt;  [83] knitr_1.23                    bit64_0.9-7                  
#&gt;  [85] fs_1.3.1                      fitdistrplus_1.0-14          
#&gt;  [87] caTools_1.17.1.2              purrr_0.3.2                  
#&gt;  [89] RANN_2.6.1                    pbapply_1.4-0                
#&gt;  [91] future_1.13.0                 nlme_3.1-140                 
#&gt;  [93] whisker_0.3-2                 mime_0.7                     
#&gt;  [95] R.oo_1.22.0                   biomaRt_2.41.3               
#&gt;  [97] compiler_3.6.0                png_0.1-7                    
#&gt;  [99] plotly_4.9.0                  curl_3.3                     
#&gt; [101] interactiveDisplayBase_1.23.0 lsei_1.2-0                   
#&gt; [103] tibble_2.1.3                  stringi_1.4.3                
#&gt; [105] lattice_0.20-38               pillar_1.4.1                 
#&gt; [107] BiocManager_1.30.4            Rdpack_0.11-0                
#&gt; [109] lmtest_0.9-37                 cowplot_0.9.4                
#&gt; [111] bitops_1.0-6                  irlba_2.3.3                  
#&gt; [113] gbRd_0.4-11                   httpuv_1.5.1                 
#&gt; [115] rtracklayer_1.45.1            R6_2.4.0                     
#&gt; [117] promises_1.0.1                gridExtra_2.3                
#&gt; [119] KernSmooth_2.23-15            codetools_0.2-16             
#&gt; [121] MASS_7.3-51.4                 gtools_3.8.1                 
#&gt; [123] assertthat_0.2.1              rhdf5_2.29.0                 
#&gt; [125] rprojroot_1.3-2               withr_2.1.2                  
#&gt; [127] sctransform_0.2.0             GenomicAlignments_1.21.3     
#&gt; [129] Rsamtools_2.1.2               GenomeInfoDbData_1.2.1       
#&gt; [131] hms_0.4.2                     grid_3.6.0                   
#&gt; [133] tidyr_0.8.3                   rmarkdown_1.13               
#&gt; [135] Rtsne_0.15                    git2r_0.25.2                 
#&gt; [137] shiny_1.3.2</code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
