<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Lambda Moses" />

<meta name="date" content="2019-07-16" />

<title>Pseudotime analysis with Monocle 2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">kallisto | bustools R Notebooks</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BUStools/BUS_notebooks_R">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Pseudotime analysis with Monocle 2</h1>
<h4 class="author">Lambda Moses</h4>
<h4 class="date">2019-07-16</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-07-16
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>BUS_notebooks_R/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.4.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongobjectspresent"> <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> <strong>Environment:</strong> objects present </a>
</p>
</div>
<div id="strongEnvironmentstrongobjectspresent" class="panel-collapse collapse">
<div class="panel-body">
<p>
The global environment had objects present when the code in the R Markdown file was run. These objects can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment. Use <code>wflow_publish</code> or <code>wflow_build</code> to ensure that the code is always run in an empty environment.
</p>
<p>
The following objects were defined in the global environment when these results were created:
</p>
<table class="table table-condensed table-hover">
<thead>
<tr>
<th style="text-align:left;">
Name
</th>
<th style="text-align:left;">
Class
</th>
<th style="text-align:left;">
Size
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
data
</td>
<td style="text-align:left;">
environment
</td>
<td style="text-align:left;">
56 bytes
</td>
</tr>
<tr>
<td style="text-align:left;">
env
</td>
<td style="text-align:left;">
environment
</td>
<td style="text-align:left;">
56 bytes
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20181214code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20181214)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20181214code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20181214)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtree3cf75fe28a1852364770b5dd229a64f8be512137targetblank3cf75fea"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/BUStools/BUS_notebooks_R/tree/3cf75fe28a1852364770b5dd229a64f8be512137" target="_blank">3cf75fe</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomBUStoolsBUSnotebooksRtree3cf75fe28a1852364770b5dd229a64f8be512137targetblank3cf75fea" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    BUS_notebooks_R.Rproj
    Ignored:    data/cDNA_introns.t2g.txt
    Ignored:    data/cDNA_transcripts.to_capture.txt
    Ignored:    data/hgmm_1k_fastqs.tar
    Ignored:    data/hs_cdna.fa.gz
    Ignored:    data/introns_transcripts.to_capture.txt
    Ignored:    data/mm_cdna.fa.gz
    Ignored:    data/neuron_10k_fastqs.tar
    Ignored:    data/neuron_10k_v3_fastqs/
    Ignored:    data/whitelist_v2.txt
    Ignored:    data/whitelist_v3.txt
    Ignored:    output/hs_mm_tr_index.idx
    Ignored:    output/hs_tr_index.idx
    Ignored:    output/mm_cDNA_introns.idx
    Ignored:    output/mm_tr_index.idx
    Ignored:    output/out_hgmm1k/

Untracked files:
    Untracked:  ASpli_binFeatures.log
    Untracked:  analysis/velocity.Rmd
    Untracked:  analysis/velocity2.Rmd
    Untracked:  data/cDNA.correct_header.fa
    Untracked:  data/fastqs/
    Untracked:  data/hs_cdna.fa
    Untracked:  data/hs_cellranger_38.gtf
    Untracked:  data/introns.correct_header.fa
    Untracked:  data/mm_cdna.fa
    Untracked:  ensembl/
    Untracked:  output/cc_neuron10k_pca.rds
    Untracked:  output/cc_neuron10k_tsne.rds
    Untracked:  output/neuron10k/
    Untracked:  output/tr2g_hgmm.tsv
    Untracked:  output/tr2g_mm.tsv
    Untracked:  output/velo_neuron10k.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/BUStools/BUS_notebooks_R/blob/3cf75fe28a1852364770b5dd229a64f8be512137/analysis/monocle2.Rmd" target="_blank">3cf75fe</a>
</td>
<td>
Lambda Moses
</td>
<td>
2019-07-17
</td>
<td>
Added Monocle 2 notebook
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>In this vignette, we will process fastq files of the <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_10k_v3">10x 10k neurons from an E18 mouse</a>, and perform pseudotime analysis with Monocle 2 on the neuronal cell types. Monocle 2 is deprecated, but it can be easily installed from Bioconductor and still has a user base. At present, Monocle 3 is at beta stage of development, and it can be hard to install on Linux servers due to gdal dependency.</p>
<div id="setup" class="section level1">
<h1>Setup</h1>
<p>If you would like to rerun this notebook, you can git clone <a href="https://github.com/BUStools/BUS_notebooks_R">this repository</a> or directly download <a href="https://github.com/BUStools/BUS_notebooks_R/blob/master/analysis/monocle2.Rmd">this notebook from GitHub</a>.</p>
<div id="install-packages" class="section level2">
<h2>Install packages</h2>
<p>This notebook demonstrates the use of command line tools <code>kallisto</code> and <code>bustools</code>. Please use <code>kallisto</code> &gt;= 0.46, whose binary can be downloaded <a href="https://pachterlab.github.io/kallisto/download">here</a>. The binary of <code>bustools</code> can be found <a href="https://github.com/BUStools/bustools/releases">here</a>.</p>
<p>After you download the binary, you should decompress the file (if it is <code>tar.gz</code>) with <code>tar -xzvf file.tar.gz</code> in the <code>bash</code> terminal, and add the directory containing the binary to <code>PATH</code> by <code>export PATH=$PATH:/foo/bar</code>, where <code>/foo/bar</code> is the directory of interest. Then you can directly invoke the binary on the command line as we will do in this notebook.</p>
<p>We will be using the R packages below. <code>BUSpaRse</code> is not yet on CRAN or Bioconductor. For Mac users, see the <a href="https://github.com/BUStools/BUSpaRse">installation note for <code>BUSpaRse</code></a>. <code>BUSpaRse</code> will be used to generate the transcript to gene file for <code>bustools</code> and to read output of <code>bustools</code> into R. We will also use <code>Seurat</code> version 3 which is now on CRAN.</p>
<pre class="r"><code># Install devtools if it&#39;s not already installed
if (!require(devtools)) {
  install.packages(&quot;devtools&quot;)
}
# Install from GitHub
devtools::install_github(&quot;BUStools/BUSpaRse&quot;)</code></pre>
<p>The package <code>SingleR</code>, which will be used for cell type inference, is only available on GitHub:</p>
<pre class="r"><code>devtools::install_github(&quot;dviraran/SingleR&quot;)</code></pre>
<p>This vignette uses the version of <code>DropletUtils</code> from Bioconductor version 3.9; the version from Bioconductor 3.8 has a different user interface. If you are using a version of R older than 3.6.0 and want to rerun this vignette, then you can adapt the knee plot code to the older version of <code>DropletUtils</code> or install <a href="https://github.com/MarioniLab/DropletUtils"><code>DropletUtils</code> from GitHub</a>. The package <code>monocle</code> should also be installed from Bioconductor:</p>
<pre class="r"><code>if (!require(BiocManager)) {
  install.packages(&quot;BiocManager&quot;)
}
BiocManager::install(c(&quot;DropletUtils&quot;, &quot;monocle&quot;))</code></pre>
<p>The other R packages below are on CRAN.</p>
<pre class="r"><code>library(BUSpaRse)
library(DropletUtils)
library(monocle)
library(SingleR)
library(Matrix)
library(tidyverse)
library(ggsci)
theme_set(theme_bw())</code></pre>
</div>
<div id="download-data" class="section level2">
<h2>Download data</h2>
<p>The dataset we are using is <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_10k_v3">10x 10k neurons from an E18 mouse</a> (almost 25 GB).</p>
<pre class="r"><code># Download data
if (!file.exists(&quot;./data/neuron_10k_fastqs.tar&quot;)) {
  download.files(&quot;http://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/3.0.0/neuron_10k_v3/neuron_10k_v3_fastqs.tar&quot;, &quot;./data/neuron_10k_fastqs.tar&quot;, method = &quot;wget&quot;, quiet = TRUE)
}</code></pre>
<pre class="bash"><code>cd ./data
tar -xvf ./neuron_10k_fastqs.tar</code></pre>
</div>
</div>
<div id="generate-the-gene-count-matrix" class="section level1">
<h1>Generate the gene count matrix</h1>
<div id="build-the-kallisto-index" class="section level2">
<h2>Build the <code>kallisto</code> index</h2>
<p>Here we use <a href="https://pachterlab.github.io/kallisto/about">kallisto</a> to pseudoalign the reads to the transcriptome and then to create the <code>bus</code> file to be converted to a sparse matrix. The first step is to build an index of the mouse transcriptome. The transcriptome downloaded here are from Ensembl version 94, released in October 2018.</p>
<pre class="r"><code># Mouse transcriptome
if (!file.exists(&quot;./data/mm_cdna.fa.gz&quot;)) {
  download.file(&quot;ftp://ftp.ensembl.org/pub/release-94/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz&quot;, &quot;./data/mm_cdna.fa.gz&quot;, method = &quot;wget&quot;, quiet = TRUE)
}</code></pre>
<pre class="bash"><code>kallisto index -i ./output/mm_tr_index.idx ./data/mm_cdna.fa.gz</code></pre>
<pre class="bash"><code>cd ./data/neuron_10k_v3_fastqs
kallisto bus -i ../../output/mm_tr_index.idx -o ../../output/neuron10k -x 10xv3 -t8 \
neuron_10k_v3_S1_L002_R1_001.fastq.gz neuron_10k_v3_S1_L002_R2_001.fastq.gz \
neuron_10k_v3_S1_L001_R1_001.fastq.gz neuron_10k_v3_S1_L001_R2_001.fastq.gz</code></pre>
</div>
<div id="run-bustools" class="section level2">
<h2>Run <code>bustools</code></h2>
<div id="map-transcripts-to-genes" class="section level3">
<h3>Map transcripts to genes</h3>
<p>For the sparse matrix, most people are interested in how many UMIs per gene per cell, we here we will quantify this from the <code>bus</code> output, and to do so, we need to find which gene corresponds to each transcript. Remember in the output of <code>kallisto bus</code>, there’s the file <code>transcripts.txt</code>. Those are the transcripts in the transcriptome index.</p>
<p>Remember that we downloaded transcriptome FASTA file from Ensembl just now. In FASTA files, each entry is a sequence with a name. In Ensembl FASTA files, the sequence name has genome annotation of the corresponding sequence, so we can extract transcript IDs and corresponding gene IDs and gene names from there.</p>
<pre class="r"><code>tr2g &lt;- transcript2gene(fasta_file = &quot;./data/mm_cdna.fa.gz&quot;, 
                        kallisto_out_path = &quot;./output/neuron10k&quot;,
                        verbose = FALSE)</code></pre>
<pre class="r"><code>head(tr2g)</code></pre>
<pre><code>#&gt;              transcript                 gene gene_name
#&gt; 1: ENSMUST00000177564.1 ENSMUSG00000096176.1     Trdd2
#&gt; 2: ENSMUST00000196221.1 ENSMUSG00000096749.2     Trdd1
#&gt; 3: ENSMUST00000179664.1 ENSMUSG00000096749.2     Trdd1
#&gt; 4: ENSMUST00000178862.1 ENSMUSG00000094569.1     Trbd2
#&gt; 5: ENSMUST00000178537.1 ENSMUSG00000095668.1     Trbd1
#&gt; 6: ENSMUST00000179520.1 ENSMUSG00000094028.1   Ighd4-1</code></pre>
<p><code>bustools</code> requires <code>tr2g</code> to be written into a tab delimited file of a specific format: No headers, first column is transcript ID, and second column is the corresponding gene ID. Transcript IDs must be in the same order as in the <code>kallisto</code> index.</p>
<pre class="r"><code># Write tr2g to format required by bustools
save_tr2g_bustools(tr2g, file_save = &quot;./output/tr2g_mm.tsv&quot;)</code></pre>
<p>A whitelist that contains all the barcodes known to be present in the kit is provided by 10x and comes with CellRanger. A CellRanger installation is required, though we will not run CellRanger here.</p>
<pre class="bash"><code>cp ~/cellranger-3.0.2/cellranger-cs/3.0.2/lib/python/cellranger/barcodes/3M-february-2018.txt.gz \
./data/whitelist_v3.txt.gz
gunzip ./data/whitelist_v3.txt.gz</code></pre>
<p>Then we’re ready to make the gene count matrix. First, <code>bustools</code> runs barcode error correction on the <code>bus</code> file. Then, the corrected <code>bus</code> file is sorted by barcode, UMI, and equivalence classes. Then the UMIs are counted and the counts are collapsed into gene level. Here the <code>|</code> is pipe in bash, just like the magrittr pipe <code>%&gt;%</code> in R, that pipes the output of one command to the next.</p>
<pre class="bash"><code>mkdir ./tmp
bustools correct -w ./data/whitelist_v3.txt -p ./output/neuron10k/output.bus | \
bustools sort -T tmp/ -t4 -p - | \
bustools count -o ./output/neuron10k/genes -g ./output/tr2g_mm.tsv \
-e ./output/neuron10k/matrix.ec -t ./output/neuron10k/transcripts.txt --genecounts -
rm -r ./tmp</code></pre>
<p>The outputs are explained in the <a href="https://bustools.github.io/BUS_notebooks_R/10xv2.html">10xv2 vignette</a>.</p>
</div>
</div>
</div>
<div id="preprocessing" class="section level1">
<h1>Preprocessing</h1>
<p>Now we can load the matrix into R for analysis.</p>
<pre class="r"><code>res_mat &lt;- read_count_output(&quot;./output/neuron10k&quot;, name = &quot;genes&quot;, tcc = FALSE)</code></pre>
<div id="remove-empty-droplets" class="section level2">
<h2>Remove empty droplets</h2>
<pre class="r"><code>dim(res_mat)</code></pre>
<pre><code>#&gt; [1]   36121 1410905</code></pre>
<p>The number of genes seems reasonable. The number of barcodes is way larger than the expected ~10k.</p>
<pre class="r"><code>tot_counts &lt;- Matrix::colSums(res_mat)
summary(tot_counts)</code></pre>
<pre><code>#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
#&gt;     0.00     1.00     1.00    64.84     2.00 51473.00</code></pre>
<p>The vast majority of “cells” have only no or just a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated methods exist (e.g. see <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets"><code>emptyDrops</code> in <code>DropletUtils</code></a>), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don’t have the ground truth.</p>
<pre class="r"><code># Compute barcode rank
bc_rank &lt;- barcodeRanks(res_mat, lower = 1000)</code></pre>
<p>Here the knee plot is transposed, because this is more generalizable to multi-modal data, such that those with not only RNA-seq but also abundance of cell surface markers. In that case, we can plot number of UMIs on the x axis, number of cell surface protein tags on the y axis, and barcode rank based on both UMI and protein tag counts on the z axis; it makes more sense to make barcode rank the dependent variable. See <a href="https://liorpachter.wordpress.com/2019/06/24/rotating-the-knee-plot-and-related-yoga/">this blog post by Lior Pachter</a> for a more detailed explanation.</p>
<pre class="r"><code>tibble(total = bc_rank$total, rank = bc_rank$rank) %&gt;% 
  distinct() %&gt;% 
  ggplot(aes(total, rank)) +
  geom_line() +
  geom_vline(xintercept = metadata(bc_rank)$knee, color = &quot;blue&quot;, linetype = 2) +
  geom_vline(xintercept = metadata(bc_rank)$inflection, color = &quot;green&quot;, linetype = 2) +
  annotate(&quot;text&quot;, y = 1000, x = 1.5 * c(metadata(bc_rank)$knee, metadata(bc_rank)$inflection),
           label = c(&quot;knee&quot;, &quot;inflection&quot;), color = c(&quot;blue&quot;, &quot;green&quot;)) +
  scale_x_log10() +
  scale_y_log10() +
  labs(y = &quot;Barcode rank&quot;, x = &quot;Total UMI count&quot;)</code></pre>
<pre><code>#&gt; Warning: Transformation introduced infinite values in continuous x-axis</code></pre>
<p><img src="figure/monocle2.Rmd/knee-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Remove genes that are not detected and empty droplets
res_mat &lt;- res_mat[Matrix::rowSums(res_mat) &gt; 0, tot_counts &gt; metadata(bc_rank)$inflection]
dim(res_mat)</code></pre>
<pre><code>#&gt; [1] 23342 11031</code></pre>
<p>Now the number of cells is closer to expectation.</p>
</div>
<div id="cell-type-inference" class="section level2">
<h2>Cell type inference</h2>
<p>Monocle 2 only infers one trajectory for the entire dataset, so non-neuronal cells like endothelial cells and erythrocytes may be mistaken as highly differentiated cells from the neuronal lineage. So we will remove non-neuronal cell types. Cell types are also helpful to orient the trajectory; neuronal progenitor cells must come before neurons. Here cell type inference is done with <a href="https://github.com/dviraran/SingleR"><code>SingleR</code></a>, which compares gene expression profiles of individual cells to bulk RNA-seq data of purified known cell types.</p>
<pre class="r"><code>data(&quot;mouse.rnaseq&quot;)
# Convert from gene symbols to Ensembl gene ID
ref_use &lt;- mouse.rnaseq$data
rownames(ref_use) &lt;- tr2g$gene[match(rownames(ref_use), tr2g$gene_name)]
ref_use &lt;- ref_use[!is.na(rownames(ref_use)),]</code></pre>
<p>Then <code>SingleR</code> will assign each cell a label based on Spearman correlation with known cell types from bulk RNA-seq. These are the labels we want:</p>
<ul>
<li>NPCs: Neuronal progenitor cells</li>
<li>aNSCs: Active neuronal stem cells</li>
<li>qNSCs: Quiescent neuronal stem cells</li>
<li>Neurons</li>
</ul>
<pre class="r"><code>annots &lt;- SingleR(&quot;single&quot;, res_mat, ref_data = ref_use, types = mouse.rnaseq$types)
inds &lt;- annots$labels %in% c(&quot;NPCs&quot;, &quot;aNSCs&quot;, &quot;qNSCs&quot;, &quot;Neurons&quot;)
# Only keep these 4 cell types
cells_use &lt;- annots$cell.names[inds]
mat_neu &lt;- res_mat[,cells_use]</code></pre>
</div>
<div id="qc" class="section level2">
<h2>QC</h2>
<pre class="r"><code>df &lt;- data.frame(nGene = Matrix::colSums(mat_neu &gt; 0),
                 nUMIs = Matrix::colSums(mat_neu))</code></pre>
<pre class="r"><code>df %&gt;% 
  gather(key = &quot;variable&quot;) %&gt;% 
  ggplot(aes(variable, value)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1, width = 0.3) +
  scale_y_log10()</code></pre>
<p><img src="figure/monocle2.Rmd/vln-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot(df, aes(nUMIs, nGene)) +
  geom_point(size = 0.5, alpha = 0.3) +
  scale_x_log10() +
  scale_y_log10()</code></pre>
<p><img src="figure/monocle2.Rmd/gene_umi-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="monocle-2" class="section level1">
<h1>Monocle 2</h1>
<pre class="r"><code># Construct CellDataSet object
pd &lt;- data.frame(cell_id = cells_use, 
                 cell_type = annots$labels[inds],
                 row.names = cells_use)
pd &lt;- new(&quot;AnnotatedDataFrame&quot;, data = pd)
fd &lt;- data.frame(gene_id = rownames(mat_neu), 
                 gene_short_name = tr2g$gene_name[match(rownames(mat_neu), tr2g$gene)],
                 row.names = row.names(mat_neu))
fd &lt;- new(&quot;AnnotatedDataFrame&quot;, data = fd)
cds &lt;- newCellDataSet(mat_neu, phenoData = pd, featureData = fd)</code></pre>
<p>Size factor and dispersion will be used to normalize data and select genes for clustering.</p>
<pre class="r"><code>cds &lt;- estimateSizeFactors(cds)
cds &lt;- estimateDispersions(cds)</code></pre>
<pre><code>#&gt; Warning in log(ifelse(y == 0, 1, y/mu)): NaNs produced</code></pre>
<pre><code>#&gt; Warning: step size truncated due to divergence</code></pre>
<pre><code>#&gt; Warning in log(ifelse(y == 0, 1, y/mu)): NaNs produced</code></pre>
<pre><code>#&gt; Warning: step size truncated due to divergence</code></pre>
<pre><code>#&gt; Warning: glm.fit: algorithm did not converge</code></pre>
<pre><code>#&gt; Removing 104 outliers</code></pre>
<p>Genes that aren’t highly expressed enough will not be used for clustering, since they may not give meaningful signal and would only add noise.</p>
<pre class="r"><code>disp_table &lt;- dispersionTable(cds)
clustering_genes &lt;- subset(disp_table, mean_expression &gt;= 0.1)
cds &lt;- setOrderingFilter(cds, clustering_genes$gene_id)</code></pre>
<pre class="r"><code>cds &lt;- reduceDimension(cds, num_dim = 40, reduction_method = &#39;tSNE&#39;)
cds &lt;- clusterCells(cds, method = &quot;louvain&quot;)
plot_cell_clusters(cds, cell_size = 0.5) +
  theme(legend.position = &quot;none&quot;) +
  labs(x = &quot;tSNE1&quot;, y = &quot;tSNE2&quot;)</code></pre>
<p><img src="figure/monocle2.Rmd/tsne1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>See where the annotated cell types are</p>
<pre class="rtsne2"><code>plot_cell_clusters(cds, cell_size = 0.5, color_by = &quot;cell_type&quot;) +
  scale_color_d3(name = &quot;cell type&quot;) +
  labs(x = &quot;tSNE1&quot;, y = &quot;tSNE2&quot;) +
  theme(legend.position = &quot;right&quot;)</code></pre>
<p>Genes likely to be informative of ordering of cells along the pseudotime trajectory will be selected for pseudotime inference.</p>
<pre class="r"><code>diff_genes &lt;- differentialGeneTest(cds, fullModelFormulaStr = &quot;~ Cluster + cell_type&quot;,
                                   cores = 4)
# Use top 3000 differentially expressed genes
ordering_genes &lt;- row.names(subset(diff_genes, qval &lt; 1e-3))[order(diff_genes$qval)][1:3000]
cds &lt;- setOrderingFilter(cds, ordering_genes)</code></pre>
<p>Here Monocle 2 will first project the data to 2 dimensions with <code>DDRTree</code>, and then do trajectory inference (<code>orderCells</code>).</p>
<pre class="r"><code>cds &lt;- reduceDimension(cds, max_components = 2, method = &#39;DDRTree&#39;)
cds &lt;- orderCells(cds)</code></pre>
<p>See what the trajectory looks like. This projection is <code>DDRTree</code>.</p>
<pre class="r"><code>plot_cell_trajectory(cds, color_by = &quot;cell_type&quot;, cell_size = 1) +
  scale_color_d3(name = &quot;cell type&quot;)</code></pre>
<p><img src="figure/monocle2.Rmd/traj1-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>In the <a href="https://www.biorxiv.org/content/10.1101/673285v1">kallisto | bustools paper</a>, I used <code>slingshot</code> for pseudotime analysis (Supplementary Figure 6.5) of this dataset, and found two neuronal end points. The result from Monocle 2 here also shows two main branches. Also, as expected, the stem cells are at the very beginning of the trajectory.</p>
<pre class="r"><code>plot_cell_trajectory(cds, color_by = &quot;Pseudotime&quot;, cell_size = 1) +
  scale_color_viridis_c()</code></pre>
<p><img src="figure/monocle2.Rmd/traj2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The pseudotime values are inverted.</p>
<pre class="r"><code>cds &lt;- orderCells(cds, reverse = TRUE)
plot_cell_trajectory(cds, color_by = &quot;Pseudotime&quot;, cell_size = 1) +
  scale_color_viridis_c()</code></pre>
<p><img src="figure/monocle2.Rmd/traj3-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Monocle 2 can also be used to find genes differentially expressed along the pseudotime trajectory and clusters of such genes. See <a href="https://davetang.org/muse/2017/10/01/getting-started-monocle/">David Tang’s excellent Monocle 2 tutorial</a> for how to use these functionalities.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>#&gt; R version 3.5.2 (2018-12-20)
#&gt; Platform: x86_64-redhat-linux-gnu (64-bit)
#&gt; Running under: CentOS Linux 7 (Core)
#&gt; 
#&gt; Matrix products: default
#&gt; BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
#&gt; 
#&gt; locale:
#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#&gt; 
#&gt; attached base packages:
#&gt;  [1] splines   parallel  stats4    stats     graphics  grDevices utils    
#&gt;  [8] datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt;  [1] ggsci_2.9                   forcats_0.4.0              
#&gt;  [3] stringr_1.4.0               dplyr_0.8.3                
#&gt;  [5] purrr_0.3.2                 readr_1.3.1                
#&gt;  [7] tidyr_0.8.3                 tibble_2.1.3               
#&gt;  [9] tidyverse_1.2.1             SingleR_0.2.2              
#&gt; [11] monocle_2.10.1              DDRTree_0.1.5              
#&gt; [13] irlba_2.3.3                 VGAM_1.1-1                 
#&gt; [15] ggplot2_3.2.0               Matrix_1.2-17              
#&gt; [17] DropletUtils_1.5.3          SingleCellExperiment_1.4.1 
#&gt; [19] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
#&gt; [21] BiocParallel_1.16.6         matrixStats_0.54.0         
#&gt; [23] Biobase_2.42.0              GenomicRanges_1.34.0       
#&gt; [25] GenomeInfoDb_1.18.2         IRanges_2.16.0             
#&gt; [27] S4Vectors_0.20.1            BiocGenerics_0.28.0        
#&gt; [29] BUSpaRse_0.99.13           
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;   [1] reticulate_1.12          R.utils_2.9.0           
#&gt;   [3] tidyselect_0.2.5         RSQLite_2.1.1           
#&gt;   [5] AnnotationDbi_1.44.0     htmlwidgets_1.3         
#&gt;   [7] grid_3.5.2               combinat_0.0-8          
#&gt;   [9] docopt_0.6.1             Rtsne_0.15              
#&gt;  [11] munsell_0.5.0            codetools_0.2-16        
#&gt;  [13] ica_1.0-2                future_1.13.0           
#&gt;  [15] withr_2.1.2              colorspace_1.4-1        
#&gt;  [17] fastICA_1.2-2            highr_0.8               
#&gt;  [19] knitr_1.23               rstudioapi_0.10         
#&gt;  [21] Seurat_3.0.2             ROCR_1.0-7              
#&gt;  [23] pbmcapply_1.4.1          gbRd_0.4-11             
#&gt;  [25] listenv_0.7.0            labeling_0.3            
#&gt;  [27] Rdpack_0.11-0            git2r_0.26.1            
#&gt;  [29] slam_0.1-45              GenomeInfoDbData_1.2.0  
#&gt;  [31] bit64_0.9-7              pheatmap_1.0.12         
#&gt;  [33] rhdf5_2.26.2             rprojroot_1.3-2         
#&gt;  [35] generics_0.0.2           vctrs_0.2.0             
#&gt;  [37] xfun_0.7                 doParallel_1.0.14       
#&gt;  [39] R6_2.4.0                 rsvd_1.0.1              
#&gt;  [41] locfit_1.5-9.1           bitops_1.0-6            
#&gt;  [43] assertthat_0.2.1         promises_1.0.1          
#&gt;  [45] SDMTools_1.1-221.1       scales_1.0.0            
#&gt;  [47] gtable_0.3.0             npsurv_0.4-0            
#&gt;  [49] globals_0.12.4           workflowr_1.4.0         
#&gt;  [51] rlang_0.4.0              zeallot_0.1.0           
#&gt;  [53] rtracklayer_1.42.2       lazyeval_0.2.2          
#&gt;  [55] broom_0.5.2              plyranges_1.2.0         
#&gt;  [57] modelr_0.1.4             yaml_2.2.0              
#&gt;  [59] reshape2_1.4.3           GenomicFeatures_1.34.8  
#&gt;  [61] backports_1.1.4          httpuv_1.5.1            
#&gt;  [63] tools_3.5.2              gplots_3.0.1.1          
#&gt;  [65] RColorBrewer_1.1-2       proxy_0.4-23            
#&gt;  [67] ggridges_0.5.1           Rcpp_1.0.1              
#&gt;  [69] plyr_1.8.4               progress_1.2.2          
#&gt;  [71] zlibbioc_1.28.0          RCurl_1.95-4.12         
#&gt;  [73] densityClust_0.3         prettyunits_1.0.2       
#&gt;  [75] pbapply_1.4-0            viridis_0.5.1           
#&gt;  [77] cowplot_1.0.0            zoo_1.8-6               
#&gt;  [79] haven_2.1.0              ggrepel_0.8.1           
#&gt;  [81] cluster_2.1.0            fs_1.3.1                
#&gt;  [83] magrittr_1.5             data.table_1.12.2       
#&gt;  [85] lmtest_0.9-37            RANN_2.6.1              
#&gt;  [87] whisker_0.3-2            fitdistrplus_1.0-14     
#&gt;  [89] mime_0.7                 hms_0.5.0               
#&gt;  [91] lsei_1.2-0               evaluate_0.14           
#&gt;  [93] GSVA_1.30.0              xtable_1.8-4            
#&gt;  [95] XML_3.98-1.20            readxl_1.3.1            
#&gt;  [97] sparsesvd_0.1-4          gridExtra_2.3           
#&gt;  [99] HSMMSingleCell_1.2.0     compiler_3.5.2          
#&gt; [101] biomaRt_2.38.0           KernSmooth_2.23-15      
#&gt; [103] crayon_1.3.4             R.oo_1.22.0             
#&gt; [105] htmltools_0.3.6          later_0.8.0             
#&gt; [107] geneplotter_1.60.0       RcppParallel_4.4.3      
#&gt; [109] lubridate_1.7.4          DBI_1.0.0               
#&gt; [111] MASS_7.3-51.4            cli_1.1.0               
#&gt; [113] R.methodsS3_1.7.1        gdata_2.18.0            
#&gt; [115] metap_1.1                igraph_1.2.4.1          
#&gt; [117] pkgconfig_2.0.2          GenomicAlignments_1.18.1
#&gt; [119] plotly_4.9.0             xml2_1.2.0              
#&gt; [121] foreach_1.4.4            annotate_1.60.1         
#&gt; [123] dqrng_0.2.1              XVector_0.22.0          
#&gt; [125] doFuture_0.8.0           rvest_0.3.4             
#&gt; [127] bibtex_0.4.2             digest_0.6.20           
#&gt; [129] sctransform_0.2.0        tsne_0.1-3              
#&gt; [131] graph_1.60.0             Biostrings_2.50.2       
#&gt; [133] cellranger_1.1.0         rmarkdown_1.13          
#&gt; [135] edgeR_3.24.3             GSEABase_1.44.0         
#&gt; [137] shiny_1.3.2              Rsamtools_1.34.1        
#&gt; [139] gtools_3.8.1             outliers_0.14           
#&gt; [141] nlme_3.1-140             jsonlite_1.6            
#&gt; [143] Rhdf5lib_1.4.3           viridisLite_0.3.0       
#&gt; [145] limma_3.38.3             BSgenome_1.50.0         
#&gt; [147] pillar_1.4.2             lattice_0.20-38         
#&gt; [149] httr_1.4.0               survival_2.44-1.1       
#&gt; [151] glue_1.3.1               qlcMatrix_0.9.7         
#&gt; [153] FNN_1.1.3                iterators_1.0.10        
#&gt; [155] shinythemes_1.1.2        png_0.1-7               
#&gt; [157] bit_1.1-14               stringi_1.4.3           
#&gt; [159] HDF5Array_1.10.1         blob_1.2.0              
#&gt; [161] singscore_1.2.2          caTools_1.17.1.2        
#&gt; [163] memoise_1.1.0            future.apply_1.3.0      
#&gt; [165] ape_5.3</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
