---
title: "1k PBMCs from a Healthy Donor (10x v3 chemistry)"
author: "Lambda Moses"
date: "`r Sys.Date()`"
output: 
  html_document:
    df_print: paged
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  comment = "#>"
)
knitr::opts_chunk$set(autodep = TRUE)
```

This vignette demonstrates how to manipulate `bus` format in R with `BUSpaRse`. The most recent version of `bustools` can generate gene count matrices from `bus` files more efficiently; the purpose of the separate implementation in `BUSpaRse` is for advanced users to experiment with new ways to collapse UMIs mapped to multiple genes and to adapt `bus` format to purposes other than single cell RNA-seq. This implementation is intended to facilitate exploration using R/Rcpp, which is easier to work with than C++.

# Setup
If you would like to rerun this notebook, you can git clone [this repository](https://github.com/BUStools/BUS_notebooks_R) or use [the Google Colab version of this notebook](https://github.com/pachterlab/kallistobustools/blob/master/notebooks/kb_10xv3.ipynb). 

## Install packages
We will be using the R packages below. `BUSpaRse` is not yet on CRAN or Bioconductor. For Mac users, see the [installation note for `BUSpaRse`](https://github.com/BUStools/BUSpaRse). The data, which is already in `bus` format, can be downloaded from the package `TENxBUSData`. Both `TENxBUSData` and `BUSpaRse` have been submitted to Bioconductor for review; the data for `TENxBUSData` can only be downloaded from Bioconductor devel (version 3.10), which requires R 3.6.0.

```{r, eval=FALSE}
# Install devtools if it's not already installed
if (!require(devtools)) {
  install.packages("devtools")
}
# Install from GitHub
devtools::install_github("BUStools/BUSpaRse")
devtools::install_github("BUStools/TENxBUSData")
```

The package `DropletUtils` will be used to estimate the number of real cells as opposed to empty droplets. It's on Bioconductor, and here is how it should be installed:
```{r, eval=FALSE}
if (!require(BiocManager)) {
  install.packages("BiocManager")
}
# Install Bioconductor devel
BiocManager::install(version = "devel")
BiocManager::install("DropletUtils")
```

The other R packages below are on CRAN, and can be installed with `install.packages`.

```{r, message=FALSE}
library(BUSpaRse)
library(TENxBUSData)
library(ggplot2)
library(magrittr)
library(data.table)
library(Seurat)
library(DropletUtils)
library(Matrix)
theme_set(theme_bw())
```

We will not get into the details of how to make the `bus` file with `kallisto bus` and `bustools`, as the data will be downloaded with `TENxBUSData`. `TENxBUSData` provides 5 different datasets, and we will use the PBMC 1k dataset here. The data from `TENxBUSData` contains the sorted `bus` file in text format. While the `BUSpaRse` package converts that text format `bus` file into gene count matrix, this text file can be loaded into R as a data frame for further exploration.
```{r}
fn <- TENxBUSData("./output", dataset = "pbmc1k")
```

```{r}
list.files(fn)
```

Explaining the output:

* `matrix.ec`: A text file with two columns. The first column is the 0 based index of equivalence classes. The second column is the set of transcripts (denoted by 0 based index based on order of appearance in the transcriptome fasta file) present in the corresponding equivalence class.
* `output.sorted`: The data represented in `bus` format, sorted by barcode, UMI, and equivalence class. This is a binary, so can't be read into R with functions like `read.table`.
* `output.sorted.txt`: `output.sorted` converted into text format, so can be easily read into R for exploration.
* `transcript.txt`: A text file with one column, which is the transcripts present in the data, in the same order as in the transcriptome fasta file.

# Sparse matrix
## Map transcripts to genes
For the sparse matrix, most people are interested in how many UMIs per gene per cell, we here we will quantify this from the `bus` output, and to do so, we need to find which gene corresponds to each transcript. Remember in the output of `kallisto bus`, there's the file `transcripts.txt`. Those are the transcripts in the transcriptome index. Information on which transcript corresponds to which gene can be directly retrieved from Ensembl.

```{r tr2g}
tr2g <- transcript2gene(species = "Homo sapiens", 
                        kallisto_out_path = "./output/out_pbmc1k",
                        ensembl_version = 94)
```

```{r}
head(tr2g)
```

Alternative ways of getting `tr2g` have been implemented in the `BUSpaRse` package. You may use `tr2g_ensembl` to query Ensembl with biomart to get transcript and gene IDs. If you use this method, then please make sure that the Ensembl version used in the query matches that of the transcriptome. This method is convenient for the user since you only need to input species names, but it can be slow since biomart database query can be slow. You may also use `tr2g_gtf` for GTF files and `tr2g_gff3` for GFF3 files, which are more useful for non-model organisms absent from Ensemble. After calling the `tr2g_*` family of functions, you should sort the transcripts from those functions with `sort_tr2g` so the transcripts are in the same order as those in the kallisto index. Then the function `save_tr2g_bustools` can be used to save the `tr2g` data frame to a text file in the format required by `bustools`.

## Make the sparse matrix
For 10x, we do have a file with all valid cell barcodes that comes with CellRanger. You need to install CellRanger to get this file, though you do not need to run CellRanger for this notebook. The whitelist is optional, so if you don't have one, you may skip the whitelist step and the `whitelist` argument in the `makr_sparse_matrix` function. 
```{bash}
# Copy v3 chemistry whitelist to working directory
cp ~/cellranger-3.1.0/cellranger-cs/3.1.0/lib/python/cellranger/barcodes/3M-february-2018.txt.gz \
./data/whitelist_v3.txt.gz
```

```{r}
# Read in the whitelist
whitelist_v3 <- fread("./data/whitelist_v3.txt.gz", header = FALSE)$V1
length(whitelist_v3)
```

That's an order of magnitude more than the 737K in v2 chemistry.

Now we have everything we need to make the sparse matrix. This function reads in `output.sorted.txt` line by line and processes them. It does not do barcode correction for now, so the barcode must exactly match those in the whitelist if one is provided. It took 5 to 6 minutes to construct the sparse matrix in the hgmm6k dataset, which has over 280 million lines in `output.sorted.txt`, which is over 9GB. Here the data set is smaller, and it takes less than a minute. 

Note that the arguments `est_ncells` (estimated number of cells) and `est_ngenes` (estimated number of genes) are important. With the estimate, this function reserves memory for the data to be added into, reducing the need of reallocation, which will slow the function down. Since the vast majority of "cells" you get in this sparse matrix are empty droplets rather than cells, please put at least 200 times more "cells" than you actually expect in `est_ncells`.

If you do not have a whitelist of barcodes, then it's fine; the `whitelist` argument is optional.

The function `make_sparse_matrix` can make the gene count matrix and the transcript compatibility count (TCC) matrix at the same time. For the purpose of this notebook, we only generate the gene count matrix. An upcoming notebook will demonstrate some more detailed analysis with a TCC matrix. See [Ntranos et al. 2016](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0970-8) for more information about TCC matrices.
```{r spmat}
res_mat <- make_sparse_matrix("./output/out_pbmc1k/output.sorted.txt",
                              tr2g = tr2g, est_ncells = 3e5,
                              est_ngenes = nrow(tr2g),
                              whitelist = whitelist_v3, TCC = FALSE)
```

The matrix we get has genes in rows and barcode in columns. The row names are the gene IDs (not using human readable gene names since they're not guaranteed to be unique), and the column names are cell barcodes.

# Explore the data
## Remove empty droplets
Cool, so now we have the sparse matrix. What does it look like?
```{r}
dim(res_mat)
```

That's way more cells than we expect, which is about 1000. So what's going on?

How many UMIs per barcode?
```{r}
tot_counts <- Matrix::colSums(res_mat)
summary(tot_counts)
```

The vast majority of "cells" have only a few UMI detected. Those are empty droplets. 10x claims to have cell capture rate of up to 65%, but in practice, depending on how many cells are in fact loaded, the rate can be much lower. A commonly used method to estimate the number of empty droplets is barcode ranking knee and inflection points, as those are often assumed to represent transition between two components of a distribution. While more sophisticated method exist (e.g. see [`emptyDrops` in `DropletUtils`](https://www.bioconductor.org/packages/devel/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html#detecting-empty-droplets)), for simplicity, we will use the barcode ranking method here. However, whichever way we go, we don't have the ground truth.
```{r}
# Compute barcode rank
bc_rank <- barcodeRanks(res_mat)
```

```{r rank}
qplot(bc_rank$total, bc_rank$rank, geom = "line") +
  geom_vline(xintercept = metadata(bc_rank)$knee, color = "blue", linetype = 2) +
  geom_vline(xintercept = metadata(bc_rank)$inflection, color = "green", linetype = 2) +
  annotate("text", y = 1000, x = 1.5 * c(metadata(bc_rank)$knee, metadata(bc_rank)$inflection),
           label = c("knee", "inflection"), color = c("blue", "green")) +
  scale_x_log10() +
  scale_y_log10() +
  labs(y = "Barcode rank", x = "Total UMI count")
```

The inflection point looks like a reasonable number of cells.
```{r}
# Filter the matrix
res_mat <- res_mat[, tot_counts > metadata(bc_rank)$inflection]
dim(res_mat)
```

## Dimension reduction
```{r seu}
seu <- CreateSeuratObject(res_mat, min.cells = 3) %>% 
  NormalizeData(verbose = FALSE) %>% 
  ScaleData(verbose = FALSE) %>% 
  FindVariableFeatures(verbose = FALSE)
```

See how number of total counts and number of genes expressed are distributed.
```{r vln}
VlnPlot(seu, c("nCount_RNA", "nFeature_RNA"), pt.size = 0.1)
```

Another QC plot
```{r}
ggplot(seu@meta.data, aes(nCount_RNA, nFeature_RNA)) +
  geom_point(alpha = 0.7, size = 0.5) +
  labs(x = "Total UMI counts per cell", y = "Number of genes detected")
```

```{r elbow}
seu <- RunPCA(seu, verbose = FALSE, npcs = 30)
ElbowPlot(seu, ndims = 30)
```

We can do Leiden clustering. Leiden is an improvement over Louvain that guarantees that clusters are well-connected on the k nearest neighbor graph.
```{r}
# Leiden clustering
seu <- FindNeighbors(seu)
seu <- FindClusters(seu, algorithm = 4)
```

```{r pca}
DimPlot(seu, reduction = "pca", pt.size = 0.5)
```

```{r tsne}
seu <- RunTSNE(seu, dims = 1:20, check_duplicates = FALSE)
DimPlot(seu, reduction = "tsne", pt.size = 0.5)
```
